{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b6b825d0",
   "metadata": {},
   "source": [
    "### Means Korrelieren"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "36d99c67",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0b079184",
   "metadata": {},
   "outputs": [],
   "source": [
    "import importlib\n",
    "import functions\n",
    "importlib.reload(functions)\n",
    "functions.set_user('Kaja')\n",
    "data = functions.call_data_clean(p_threshold=1.3)\n",
    "# results = functions.call_data_clean()\n",
    "# print(results['ATAC_seq'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fab13df7",
   "metadata": {},
   "outputs": [],
   "source": [
    "ATAC = data['ATAC_seq']\n",
    "RNA = data['RNA_seq']\n",
    "QC = data['QC_metrics']\n",
    "#ATAC_scores = data['ATAC_seq_only_scores']\n",
    "#ATACtest = data['test1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "531bcac0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>Summit</th>\n",
       "      <th>mm10.60way.phastCons_scores</th>\n",
       "      <th>_-log10_bestPvalue</th>\n",
       "      <th>Included.in.systematic.analysis</th>\n",
       "      <th>TSS</th>\n",
       "      <th>genes.within.100Kb</th>\n",
       "      <th>LTHSC.34-.BM</th>\n",
       "      <th>LTHSC.34+.BM</th>\n",
       "      <th>STHSC.150-.BM</th>\n",
       "      <th>...</th>\n",
       "      <th>DC.4+.Sp</th>\n",
       "      <th>DC.8+.Sp</th>\n",
       "      <th>DC.pDC.Sp</th>\n",
       "      <th>DC.103+11b+.SI</th>\n",
       "      <th>DC.103+11b-.SI</th>\n",
       "      <th>FRC.SLN</th>\n",
       "      <th>IAP.SLN</th>\n",
       "      <th>BEC.SLN</th>\n",
       "      <th>LEC.SLN</th>\n",
       "      <th>Ep.MEChi.Th</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ImmGenATAC1219.peakID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ImmGenATAC1219.peak_3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3120109</td>\n",
       "      <td>0.07</td>\n",
       "      <td>10.80</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.205769</td>\n",
       "      <td>0.074722</td>\n",
       "      <td>0.560426</td>\n",
       "      <td>...</td>\n",
       "      <td>0.545095</td>\n",
       "      <td>0.364004</td>\n",
       "      <td>1.300606</td>\n",
       "      <td>0.587422</td>\n",
       "      <td>0.081492</td>\n",
       "      <td>5.160862</td>\n",
       "      <td>2.528480</td>\n",
       "      <td>0.801883</td>\n",
       "      <td>0.670699</td>\n",
       "      <td>0.082292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ImmGenATAC1219.peak_4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3121485</td>\n",
       "      <td>0.15</td>\n",
       "      <td>3.02</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.295129</td>\n",
       "      <td>0.074722</td>\n",
       "      <td>0.081366</td>\n",
       "      <td>...</td>\n",
       "      <td>0.301232</td>\n",
       "      <td>0.980088</td>\n",
       "      <td>0.438368</td>\n",
       "      <td>0.081600</td>\n",
       "      <td>0.571458</td>\n",
       "      <td>3.071593</td>\n",
       "      <td>0.607880</td>\n",
       "      <td>0.778569</td>\n",
       "      <td>0.670699</td>\n",
       "      <td>0.082292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ImmGenATAC1219.peak_5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3372787</td>\n",
       "      <td>0.03</td>\n",
       "      <td>1.31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.295129</td>\n",
       "      <td>0.074722</td>\n",
       "      <td>0.081366</td>\n",
       "      <td>...</td>\n",
       "      <td>0.301232</td>\n",
       "      <td>0.074707</td>\n",
       "      <td>0.438368</td>\n",
       "      <td>0.961199</td>\n",
       "      <td>0.344177</td>\n",
       "      <td>0.589547</td>\n",
       "      <td>0.486492</td>\n",
       "      <td>0.801883</td>\n",
       "      <td>0.938263</td>\n",
       "      <td>1.766096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ImmGenATAC1219.peak_6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3399217</td>\n",
       "      <td>0.06</td>\n",
       "      <td>2.39</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.205769</td>\n",
       "      <td>0.904419</td>\n",
       "      <td>0.982424</td>\n",
       "      <td>...</td>\n",
       "      <td>0.771585</td>\n",
       "      <td>0.206800</td>\n",
       "      <td>0.168304</td>\n",
       "      <td>0.556619</td>\n",
       "      <td>0.081492</td>\n",
       "      <td>0.368259</td>\n",
       "      <td>0.806198</td>\n",
       "      <td>0.587879</td>\n",
       "      <td>1.396850</td>\n",
       "      <td>2.548602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ImmGenATAC1219.peak_7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3400115</td>\n",
       "      <td>0.44</td>\n",
       "      <td>2.57</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.295129</td>\n",
       "      <td>0.074722</td>\n",
       "      <td>0.081366</td>\n",
       "      <td>...</td>\n",
       "      <td>0.545095</td>\n",
       "      <td>1.018444</td>\n",
       "      <td>0.397707</td>\n",
       "      <td>0.081600</td>\n",
       "      <td>0.794370</td>\n",
       "      <td>1.272359</td>\n",
       "      <td>0.486492</td>\n",
       "      <td>1.192532</td>\n",
       "      <td>1.221350</td>\n",
       "      <td>2.782528</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 97 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      chrom   Summit  mm10.60way.phastCons_scores  \\\n",
       "ImmGenATAC1219.peakID                                               \n",
       "ImmGenATAC1219.peak_3  chr1  3120109                         0.07   \n",
       "ImmGenATAC1219.peak_4  chr1  3121485                         0.15   \n",
       "ImmGenATAC1219.peak_5  chr1  3372787                         0.03   \n",
       "ImmGenATAC1219.peak_6  chr1  3399217                         0.06   \n",
       "ImmGenATAC1219.peak_7  chr1  3400115                         0.44   \n",
       "\n",
       "                       _-log10_bestPvalue  Included.in.systematic.analysis  \\\n",
       "ImmGenATAC1219.peakID                                                        \n",
       "ImmGenATAC1219.peak_3               10.80                              1.0   \n",
       "ImmGenATAC1219.peak_4                3.02                              1.0   \n",
       "ImmGenATAC1219.peak_5                1.31                              NaN   \n",
       "ImmGenATAC1219.peak_6                2.39                              1.0   \n",
       "ImmGenATAC1219.peak_7                2.57                              1.0   \n",
       "\n",
       "                       TSS genes.within.100Kb  LTHSC.34-.BM  LTHSC.34+.BM  \\\n",
       "ImmGenATAC1219.peakID                                                       \n",
       "ImmGenATAC1219.peak_3  NaN                NaN      1.205769      0.074722   \n",
       "ImmGenATAC1219.peak_4  NaN                NaN      0.295129      0.074722   \n",
       "ImmGenATAC1219.peak_5  NaN                NaN      0.295129      0.074722   \n",
       "ImmGenATAC1219.peak_6  NaN                NaN      1.205769      0.904419   \n",
       "ImmGenATAC1219.peak_7  NaN                NaN      0.295129      0.074722   \n",
       "\n",
       "                       STHSC.150-.BM  ...  DC.4+.Sp  DC.8+.Sp  DC.pDC.Sp  \\\n",
       "ImmGenATAC1219.peakID                 ...                                  \n",
       "ImmGenATAC1219.peak_3       0.560426  ...  0.545095  0.364004   1.300606   \n",
       "ImmGenATAC1219.peak_4       0.081366  ...  0.301232  0.980088   0.438368   \n",
       "ImmGenATAC1219.peak_5       0.081366  ...  0.301232  0.074707   0.438368   \n",
       "ImmGenATAC1219.peak_6       0.982424  ...  0.771585  0.206800   0.168304   \n",
       "ImmGenATAC1219.peak_7       0.081366  ...  0.545095  1.018444   0.397707   \n",
       "\n",
       "                       DC.103+11b+.SI  DC.103+11b-.SI   FRC.SLN   IAP.SLN  \\\n",
       "ImmGenATAC1219.peakID                                                       \n",
       "ImmGenATAC1219.peak_3        0.587422        0.081492  5.160862  2.528480   \n",
       "ImmGenATAC1219.peak_4        0.081600        0.571458  3.071593  0.607880   \n",
       "ImmGenATAC1219.peak_5        0.961199        0.344177  0.589547  0.486492   \n",
       "ImmGenATAC1219.peak_6        0.556619        0.081492  0.368259  0.806198   \n",
       "ImmGenATAC1219.peak_7        0.081600        0.794370  1.272359  0.486492   \n",
       "\n",
       "                        BEC.SLN   LEC.SLN  Ep.MEChi.Th  \n",
       "ImmGenATAC1219.peakID                                   \n",
       "ImmGenATAC1219.peak_3  0.801883  0.670699     0.082292  \n",
       "ImmGenATAC1219.peak_4  0.778569  0.670699     0.082292  \n",
       "ImmGenATAC1219.peak_5  0.801883  0.938263     1.766096  \n",
       "ImmGenATAC1219.peak_6  0.587879  1.396850     2.548602  \n",
       "ImmGenATAC1219.peak_7  1.192532  1.221350     2.782528  \n",
       "\n",
       "[5 rows x 97 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ATAC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8b3720c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CellType</th>\n",
       "      <th>ImmGenLab</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>CellFamily</th>\n",
       "      <th>Organ</th>\n",
       "      <th>SortingMarkers</th>\n",
       "      <th>InputCellNumber</th>\n",
       "      <th>PF.reads</th>\n",
       "      <th>%chrM.mapped</th>\n",
       "      <th>Paired.read.after.removing.PCR.duplication</th>\n",
       "      <th>%fragment.1Kb_TSS</th>\n",
       "      <th>Replicate.cor</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleName</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LTHSC.34-.BM#1</th>\n",
       "      <td>LTHSC.34-.BM</td>\n",
       "      <td>Wagers</td>\n",
       "      <td>Stem&amp;Prog</td>\n",
       "      <td>Stem&amp;Prog Cell</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>Lin-Sca1+ckit+CD135-CD150+CD48-CD34-</td>\n",
       "      <td>677</td>\n",
       "      <td>22287984</td>\n",
       "      <td>4.58</td>\n",
       "      <td>2470102</td>\n",
       "      <td>10.60</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LTHSC.34+.BM#1</th>\n",
       "      <td>LTHSC.34+.BM</td>\n",
       "      <td>Wagers</td>\n",
       "      <td>Stem&amp;Prog</td>\n",
       "      <td>Stem&amp;Prog Cell</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>Lin-Sca1+ckit+CD135-CD150+CD48-CD34+</td>\n",
       "      <td>2483</td>\n",
       "      <td>28588536</td>\n",
       "      <td>3.17</td>\n",
       "      <td>7675501</td>\n",
       "      <td>7.70</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STHSC.150-.BM#1</th>\n",
       "      <td>STHSC.150-.BM</td>\n",
       "      <td>Wagers</td>\n",
       "      <td>Stem&amp;Prog</td>\n",
       "      <td>Stem&amp;Prog Cell</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>Lin-Sca1+ckit+CD135-CD150-CD48-</td>\n",
       "      <td>3660</td>\n",
       "      <td>22170346</td>\n",
       "      <td>3.25</td>\n",
       "      <td>6060045</td>\n",
       "      <td>11.13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MPP4.135+.BM#1</th>\n",
       "      <td>MPP4.135+.BM</td>\n",
       "      <td>Wagers</td>\n",
       "      <td>Stem&amp;Prog</td>\n",
       "      <td>Stem&amp;Prog Cell</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>Lin-Sca1+ckit+CD135+</td>\n",
       "      <td>10000</td>\n",
       "      <td>24003590</td>\n",
       "      <td>13.83</td>\n",
       "      <td>6128252</td>\n",
       "      <td>18.96</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>proB.CLP.BM#1</th>\n",
       "      <td>proB.CLP.BM</td>\n",
       "      <td>Hardy</td>\n",
       "      <td>B</td>\n",
       "      <td>B Cell</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>CD19-IgM-CD43+CD24-CD45R-CD93+CD117+IL7Ra+PI-D...</td>\n",
       "      <td>10000</td>\n",
       "      <td>23578208</td>\n",
       "      <td>7.67</td>\n",
       "      <td>6588564</td>\n",
       "      <td>13.21</td>\n",
       "      <td>0.868</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      CellType ImmGenLab    Lineage      CellFamily  \\\n",
       "SampleName                                                            \n",
       "LTHSC.34-.BM#1    LTHSC.34-.BM    Wagers  Stem&Prog  Stem&Prog Cell   \n",
       "LTHSC.34+.BM#1    LTHSC.34+.BM    Wagers  Stem&Prog  Stem&Prog Cell   \n",
       "STHSC.150-.BM#1  STHSC.150-.BM    Wagers  Stem&Prog  Stem&Prog Cell   \n",
       "MPP4.135+.BM#1    MPP4.135+.BM    Wagers  Stem&Prog  Stem&Prog Cell   \n",
       "proB.CLP.BM#1      proB.CLP.BM     Hardy          B          B Cell   \n",
       "\n",
       "                       Organ  \\\n",
       "SampleName                     \n",
       "LTHSC.34-.BM#1   Bone Marrow   \n",
       "LTHSC.34+.BM#1   Bone Marrow   \n",
       "STHSC.150-.BM#1  Bone Marrow   \n",
       "MPP4.135+.BM#1   Bone Marrow   \n",
       "proB.CLP.BM#1    Bone Marrow   \n",
       "\n",
       "                                                    SortingMarkers  \\\n",
       "SampleName                                                           \n",
       "LTHSC.34-.BM#1                Lin-Sca1+ckit+CD135-CD150+CD48-CD34-   \n",
       "LTHSC.34+.BM#1                Lin-Sca1+ckit+CD135-CD150+CD48-CD34+   \n",
       "STHSC.150-.BM#1                    Lin-Sca1+ckit+CD135-CD150-CD48-   \n",
       "MPP4.135+.BM#1                                Lin-Sca1+ckit+CD135+   \n",
       "proB.CLP.BM#1    CD19-IgM-CD43+CD24-CD45R-CD93+CD117+IL7Ra+PI-D...   \n",
       "\n",
       "                 InputCellNumber  PF.reads  %chrM.mapped  \\\n",
       "SampleName                                                 \n",
       "LTHSC.34-.BM#1               677  22287984          4.58   \n",
       "LTHSC.34+.BM#1              2483  28588536          3.17   \n",
       "STHSC.150-.BM#1             3660  22170346          3.25   \n",
       "MPP4.135+.BM#1             10000  24003590         13.83   \n",
       "proB.CLP.BM#1              10000  23578208          7.67   \n",
       "\n",
       "                 Paired.read.after.removing.PCR.duplication  \\\n",
       "SampleName                                                    \n",
       "LTHSC.34-.BM#1                                      2470102   \n",
       "LTHSC.34+.BM#1                                      7675501   \n",
       "STHSC.150-.BM#1                                     6060045   \n",
       "MPP4.135+.BM#1                                      6128252   \n",
       "proB.CLP.BM#1                                       6588564   \n",
       "\n",
       "                 %fragment.1Kb_TSS  Replicate.cor  \n",
       "SampleName                                         \n",
       "LTHSC.34-.BM#1               10.60            NaN  \n",
       "LTHSC.34+.BM#1                7.70            NaN  \n",
       "STHSC.150-.BM#1              11.13            NaN  \n",
       "MPP4.135+.BM#1               18.96            NaN  \n",
       "proB.CLP.BM#1                13.21          0.868  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "QC.head()\n",
    "\n",
    "#fehler: peak 1 und2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "86c86112",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CellType</th>\n",
       "      <th>ImmGenLab</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>CellFamily</th>\n",
       "      <th>Organ</th>\n",
       "      <th>SortingMarkers</th>\n",
       "      <th>InputCellNumber</th>\n",
       "      <th>PF.reads</th>\n",
       "      <th>%chrM.mapped</th>\n",
       "      <th>Paired.read.after.removing.PCR.duplication</th>\n",
       "      <th>%fragment.1Kb_TSS</th>\n",
       "      <th>Replicate.cor</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleName</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LTHSC.34-.BM#1</th>\n",
       "      <td>LTHSC.34-.BM</td>\n",
       "      <td>Wagers</td>\n",
       "      <td>Stem&amp;Prog</td>\n",
       "      <td>Stem&amp;Prog Cell</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>Lin-Sca1+ckit+CD135-CD150+CD48-CD34-</td>\n",
       "      <td>677</td>\n",
       "      <td>22287984</td>\n",
       "      <td>4.58</td>\n",
       "      <td>2470102</td>\n",
       "      <td>10.60</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LTHSC.34+.BM#1</th>\n",
       "      <td>LTHSC.34+.BM</td>\n",
       "      <td>Wagers</td>\n",
       "      <td>Stem&amp;Prog</td>\n",
       "      <td>Stem&amp;Prog Cell</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>Lin-Sca1+ckit+CD135-CD150+CD48-CD34+</td>\n",
       "      <td>2483</td>\n",
       "      <td>28588536</td>\n",
       "      <td>3.17</td>\n",
       "      <td>7675501</td>\n",
       "      <td>7.70</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STHSC.150-.BM#1</th>\n",
       "      <td>STHSC.150-.BM</td>\n",
       "      <td>Wagers</td>\n",
       "      <td>Stem&amp;Prog</td>\n",
       "      <td>Stem&amp;Prog Cell</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>Lin-Sca1+ckit+CD135-CD150-CD48-</td>\n",
       "      <td>3660</td>\n",
       "      <td>22170346</td>\n",
       "      <td>3.25</td>\n",
       "      <td>6060045</td>\n",
       "      <td>11.13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MPP4.135+.BM#1</th>\n",
       "      <td>MPP4.135+.BM</td>\n",
       "      <td>Wagers</td>\n",
       "      <td>Stem&amp;Prog</td>\n",
       "      <td>Stem&amp;Prog Cell</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>Lin-Sca1+ckit+CD135+</td>\n",
       "      <td>10000</td>\n",
       "      <td>24003590</td>\n",
       "      <td>13.83</td>\n",
       "      <td>6128252</td>\n",
       "      <td>18.96</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>proB.CLP.BM#1</th>\n",
       "      <td>proB.CLP.BM</td>\n",
       "      <td>Hardy</td>\n",
       "      <td>B</td>\n",
       "      <td>B Cell</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>CD19-IgM-CD43+CD24-CD45R-CD93+CD117+IL7Ra+PI-D...</td>\n",
       "      <td>10000</td>\n",
       "      <td>23578208</td>\n",
       "      <td>7.67</td>\n",
       "      <td>6588564</td>\n",
       "      <td>13.21</td>\n",
       "      <td>0.868</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      CellType ImmGenLab    Lineage      CellFamily  \\\n",
       "SampleName                                                            \n",
       "LTHSC.34-.BM#1    LTHSC.34-.BM    Wagers  Stem&Prog  Stem&Prog Cell   \n",
       "LTHSC.34+.BM#1    LTHSC.34+.BM    Wagers  Stem&Prog  Stem&Prog Cell   \n",
       "STHSC.150-.BM#1  STHSC.150-.BM    Wagers  Stem&Prog  Stem&Prog Cell   \n",
       "MPP4.135+.BM#1    MPP4.135+.BM    Wagers  Stem&Prog  Stem&Prog Cell   \n",
       "proB.CLP.BM#1      proB.CLP.BM     Hardy          B          B Cell   \n",
       "\n",
       "                       Organ  \\\n",
       "SampleName                     \n",
       "LTHSC.34-.BM#1   Bone Marrow   \n",
       "LTHSC.34+.BM#1   Bone Marrow   \n",
       "STHSC.150-.BM#1  Bone Marrow   \n",
       "MPP4.135+.BM#1   Bone Marrow   \n",
       "proB.CLP.BM#1    Bone Marrow   \n",
       "\n",
       "                                                    SortingMarkers  \\\n",
       "SampleName                                                           \n",
       "LTHSC.34-.BM#1                Lin-Sca1+ckit+CD135-CD150+CD48-CD34-   \n",
       "LTHSC.34+.BM#1                Lin-Sca1+ckit+CD135-CD150+CD48-CD34+   \n",
       "STHSC.150-.BM#1                    Lin-Sca1+ckit+CD135-CD150-CD48-   \n",
       "MPP4.135+.BM#1                                Lin-Sca1+ckit+CD135+   \n",
       "proB.CLP.BM#1    CD19-IgM-CD43+CD24-CD45R-CD93+CD117+IL7Ra+PI-D...   \n",
       "\n",
       "                 InputCellNumber  PF.reads  %chrM.mapped  \\\n",
       "SampleName                                                 \n",
       "LTHSC.34-.BM#1               677  22287984          4.58   \n",
       "LTHSC.34+.BM#1              2483  28588536          3.17   \n",
       "STHSC.150-.BM#1             3660  22170346          3.25   \n",
       "MPP4.135+.BM#1             10000  24003590         13.83   \n",
       "proB.CLP.BM#1              10000  23578208          7.67   \n",
       "\n",
       "                 Paired.read.after.removing.PCR.duplication  \\\n",
       "SampleName                                                    \n",
       "LTHSC.34-.BM#1                                      2470102   \n",
       "LTHSC.34+.BM#1                                      7675501   \n",
       "STHSC.150-.BM#1                                     6060045   \n",
       "MPP4.135+.BM#1                                      6128252   \n",
       "proB.CLP.BM#1                                       6588564   \n",
       "\n",
       "                 %fragment.1Kb_TSS  Replicate.cor  \n",
       "SampleName                                         \n",
       "LTHSC.34-.BM#1               10.60            NaN  \n",
       "LTHSC.34+.BM#1                7.70            NaN  \n",
       "STHSC.150-.BM#1              11.13            NaN  \n",
       "MPP4.135+.BM#1               18.96            NaN  \n",
       "proB.CLP.BM#1                13.21          0.868  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ATAC_seq_path = \"D:\\\\Uni\\\\4. Semester\\\\Bioinfo\\\\datasets\\\\processed atac seq data and called peaks.csv\"\n",
    "RNA_seq_path = \"D:\\\\Uni\\\\4. Semester\\\\Bioinfo\\\\datasets\\\\processed rna seq data.csv\"\n",
    "Transcription_exons_path = \"D:\\\\Uni\\\\4. Semester\\\\Bioinfo\\\\datasets\\\\refFlat.txt\"\n",
    "Cell_population_qc_path = \"D:\\\\Uni\\\\4. Semester\\\\Bioinfo\\\\datasets\\\\summary of immune cell populationsprofiled by atac.xlsx\"\n",
    "Voluntary_path = \"D:\\\\Uni\\\\4. Semester\\\\Bioinfo\\\\datasets\\\\summary of immune cell populationsprofiled by atac.xlsx\"\n",
    "QC_metrics = pd.read_excel(Cell_population_qc_path, header=0, index_col=0)\n",
    "QC = data['QC_metrics']\n",
    "#QC=QC.iloc[:, 0:]\n",
    "QC.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6f6373b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['LTHSC.34-.BM#1', 'LTHSC.34+.BM#1', 'STHSC.150-.BM#1', 'MPP4.135+.BM#1',\n",
      "       'proB.CLP.BM#1', 'proB.CLP.BM#2', 'proB.FrA.BM#1', 'proB.FrA.BM#2',\n",
      "       'proB.FrBC.BM#1', 'proB.FrBC.BM#2',\n",
      "       ...\n",
      "       'FRC.CD140a+.Madcam-.CD35-.SLN#1', 'FRC.CD140a+.Madcam-.CD35-.SLN#2',\n",
      "       'IAP.SLN#1', 'IAP.SLN#2', 'BEC.SLN#1', 'BEC.SLN#2', 'LEC.SLN#1',\n",
      "       'LEC.SLN#2', 'Ep.MEChi.Th#1', 'Ep.MEChi.Th#2'],\n",
      "      dtype='object', name='SampleName', length=181)\n",
      "Index(['CellType', 'ImmGenLab', 'Lineage', 'CellFamily', 'Organ',\n",
      "       'SortingMarkers', 'InputCellNumber', 'PF.reads', '%chrM.mapped',\n",
      "       'Paired.read.after.removing.PCR.duplication', '%fragment.1Kb_TSS',\n",
      "       'Replicate.cor'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "QC.head()\n",
    "#erste spalte löschen\n",
    "#QC=QC.iloc[:, 0:]\n",
    "\n",
    "#Index festlegen\n",
    "#QC.set_index('CellType', inplace=True)\n",
    "# QC.head(20)\n",
    "print(QC.index)\n",
    "print(QC.columns)\n",
    "#QC.index=QC.iloc[:, 0].str.strip().str.lower().str.replace(\" \", \"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "97fbcb44",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pearson_correlation_QC_metrics (plot_type, atac_data, qc_data, stat_type='mean', qc_metric_col=None):\n",
    "\n",
    "    '''\n",
    "    df1: first Dataframe (ATAC)\n",
    "    df2: second Dataframe (QC)\n",
    "    qc_metrics_col: welche qc metric soll korreliert werden, spaltenname\n",
    "    plot_type: scatterplot/heatmap\n",
    "    '''\n",
    "    atac_data = atac_data.select_dtypes(include='number')\n",
    "\n",
    "    atac_data.columns = atac_data.columns.astype(str).str.lower().str.replace(\" \", \"\")\n",
    "    \n",
    "    qc_data=qc_data.iloc[:, 1:]\n",
    "    qc_data.index=qc_data.iloc[:, 0:]\n",
    "    qc_data.index = qc_data.index.astype(str).str.lower().str.replace(\" \", \"\")\n",
    "\n",
    "    # Einheitliche Formatierung der Zelltypen\n",
    "    atac_data.columns = atac_data.columns.str.lower().str.replace(\" \", \"\")\n",
    "    qc_data.index = qc_data.index.str.lower().str.replace(\" \", \"\")\n",
    "\n",
    "    # Gemeinsame Zelltypen identifizieren\n",
    "    gemeinsame_zelltypen = set(atac_data.columns).intersection(qc_data.index)\n",
    "    print(f\"Anzahl gemeinsamer Zelltypen: {len(gemeinsame_zelltypen)}\")\n",
    "    print(\"Gemeinsame Zelltypen (Beispiel):\", list(gemeinsame_zelltypen)[:10])\n",
    "\n",
    "    # Zelltypen, die nur in einem der DataFrames vorkommen\n",
    "    print(\"\\nZelltypen nur in ATAC:\", list(set(atac_data.columns) - set(qc_data.index))[:10])\n",
    "    print(\"Zelltypen nur in QC:\", list(set(qc_data.index) - set(atac_data.columns))[:10])\n",
    "\n",
    "    # QC = QC.drop(QC.columns[0:2], axis=1)\n",
    "    # QC.index = QC.iloc[:, 1].str.lower().str.replace(\" \", \"\")\n",
    "\n",
    "    if stat_type=='mean':\n",
    "        atac_stat=atac_data.mean(axis=0).rename('ATAC_stat')\n",
    "    elif stat_type=='median':\n",
    "        atac_stat=atac_data.median(axis=0).rename('ATAC_stat')\n",
    "    elif stat_type=='std':\n",
    "        atac_stat=atac_data.std(axis=0).rename('ATAC_stat')\n",
    "\n",
    "    if qc_metric_col not in qc_data.columns:\n",
    "        raise ValueError(f\"QC-Metrik '{qc_metric_col}' nicht gefunden.\")\n",
    "    \n",
    "    qc_metric = qc_data[qc_metric_col]\n",
    "\n",
    "    df=pd.concat([atac_stat, qc_metric], axis=1, join='inner')\n",
    "    df.columns=['ATAC_stat', qc_metric_col]\n",
    "\n",
    "    if len(df)<2:\n",
    "       raise ValueError(\"Nicht genügend überlappende Zelltypen für Korrelation\")\n",
    "\n",
    "\n",
    "    # qc_metric=qc_data.set_index('CellType') [[qc_metric_col]]\n",
    "    # merged=pd.merge(atac_means, qc_metric, left_index=True, right_index=True)\n",
    "    r, p=pearsonr(df['ATAC_stat'], df[qc_metric_col])\n",
    "    print(f\"pearson-correlation ({stat_type} vs {qc_metric_col}): r= {r:.4f}, p-Wert: {p:.4e}\")\n",
    "     \n",
    "            #Scatterplot\n",
    "    if plot_type.lower()=='scatterplot':\n",
    "        plt.figure(figsize=(6,4))\n",
    "        sns.scatterplot(data=df, x='ATAC_mean', y=qc_metric_col)\n",
    "        slope, intercept, *_ = linregress(df['ATAC_mean'], df[qc_metric_col])\n",
    "        x_vals = pd.Series(sorted(df['ATAC_mean']))\n",
    "        y_vals = slope * x_vals + intercept\n",
    "        plt.plot(x_vals, y_vals, color='purple', label='Regression')\n",
    "        plt.xlabel('ATAC seq mean')\n",
    "        plt.ylabel(qc_metric_col)\n",
    "        plt.title(f'ATAC seq-{qc_metric_col}')\n",
    "        plt.legend()\n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "            \n",
    "            #heatmap\n",
    "    elif plot_type.lower()=='heatmap':\n",
    "        plt.figure(figsize=(3,3))\n",
    "        sns.heatmap([[r]], annot=True, cmap='coolwarm', cbar=True, xticklabels=[qc_metric_col], yticklabels=[\"ATAC_mean\"])\n",
    "        plt.title('Pearson correlation')\n",
    "        plt.tight_layout()\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0c16913c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ImmGenATAC1219.peakID</th>\n",
       "      <th>ImmGenATAC1219.peak_3</th>\n",
       "      <th>ImmGenATAC1219.peak_4</th>\n",
       "      <th>ImmGenATAC1219.peak_5</th>\n",
       "      <th>ImmGenATAC1219.peak_6</th>\n",
       "      <th>ImmGenATAC1219.peak_7</th>\n",
       "      <th>ImmGenATAC1219.peak_8</th>\n",
       "      <th>ImmGenATAC1219.peak_9</th>\n",
       "      <th>ImmGenATAC1219.peak_10</th>\n",
       "      <th>ImmGenATAC1219.peak_12</th>\n",
       "      <th>ImmGenATAC1219.peak_14</th>\n",
       "      <th>...</th>\n",
       "      <th>ImmGenATAC1219.peak_512584</th>\n",
       "      <th>ImmGenATAC1219.peak_512585</th>\n",
       "      <th>ImmGenATAC1219.peak_512586</th>\n",
       "      <th>ImmGenATAC1219.peak_512587</th>\n",
       "      <th>ImmGenATAC1219.peak_512588</th>\n",
       "      <th>ImmGenATAC1219.peak_512589</th>\n",
       "      <th>ImmGenATAC1219.peak_512590</th>\n",
       "      <th>ImmGenATAC1219.peak_512591</th>\n",
       "      <th>ImmGenATAC1219.peak_512592</th>\n",
       "      <th>ImmGenATAC1219.peak_512595</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LTHSC.34-.BM</th>\n",
       "      <td>1.748461</td>\n",
       "      <td>0.495695</td>\n",
       "      <td>0.495695</td>\n",
       "      <td>1.748461</td>\n",
       "      <td>0.495695</td>\n",
       "      <td>1.748461</td>\n",
       "      <td>2.424922</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>1.748461</td>\n",
       "      <td>0.495695</td>\n",
       "      <td>...</td>\n",
       "      <td>1.748461</td>\n",
       "      <td>1.748461</td>\n",
       "      <td>2.424922</td>\n",
       "      <td>0.495695</td>\n",
       "      <td>0.495695</td>\n",
       "      <td>0.495695</td>\n",
       "      <td>1.748461</td>\n",
       "      <td>2.424922</td>\n",
       "      <td>0.495695</td>\n",
       "      <td>0.495695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LTHSC.34+.BM</th>\n",
       "      <td>0.137504</td>\n",
       "      <td>0.137504</td>\n",
       "      <td>0.137504</td>\n",
       "      <td>1.400538</td>\n",
       "      <td>0.137504</td>\n",
       "      <td>0.773996</td>\n",
       "      <td>0.137504</td>\n",
       "      <td>0.137504</td>\n",
       "      <td>0.137504</td>\n",
       "      <td>0.137504</td>\n",
       "      <td>...</td>\n",
       "      <td>3.072106</td>\n",
       "      <td>1.835924</td>\n",
       "      <td>3.844988</td>\n",
       "      <td>1.835924</td>\n",
       "      <td>3.493135</td>\n",
       "      <td>3.072106</td>\n",
       "      <td>3.291309</td>\n",
       "      <td>3.291309</td>\n",
       "      <td>3.072106</td>\n",
       "      <td>0.773996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STHSC.150-.BM</th>\n",
       "      <td>0.925999</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>1.510962</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>0.925999</td>\n",
       "      <td>0.925999</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>0.925999</td>\n",
       "      <td>...</td>\n",
       "      <td>1.883621</td>\n",
       "      <td>1.510962</td>\n",
       "      <td>3.761285</td>\n",
       "      <td>1.883621</td>\n",
       "      <td>2.176323</td>\n",
       "      <td>1.510962</td>\n",
       "      <td>0.925999</td>\n",
       "      <td>2.176323</td>\n",
       "      <td>2.176323</td>\n",
       "      <td>0.925999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MPP4.135+.BM</th>\n",
       "      <td>0.150560</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>1.367371</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>0.871844</td>\n",
       "      <td>0.871844</td>\n",
       "      <td>...</td>\n",
       "      <td>2.965323</td>\n",
       "      <td>1.655352</td>\n",
       "      <td>3.628774</td>\n",
       "      <td>1.839960</td>\n",
       "      <td>1.655352</td>\n",
       "      <td>2.166715</td>\n",
       "      <td>1.655352</td>\n",
       "      <td>1.839960</td>\n",
       "      <td>2.014355</td>\n",
       "      <td>2.166715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>proB.CLP.BM</th>\n",
       "      <td>0.555816</td>\n",
       "      <td>0.839960</td>\n",
       "      <td>0.555816</td>\n",
       "      <td>1.084064</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>1.361768</td>\n",
       "      <td>0.505891</td>\n",
       "      <td>0.555816</td>\n",
       "      <td>1.555816</td>\n",
       "      <td>0.555816</td>\n",
       "      <td>...</td>\n",
       "      <td>4.192983</td>\n",
       "      <td>3.768714</td>\n",
       "      <td>4.397803</td>\n",
       "      <td>2.480265</td>\n",
       "      <td>1.959770</td>\n",
       "      <td>2.757023</td>\n",
       "      <td>2.087463</td>\n",
       "      <td>2.381283</td>\n",
       "      <td>2.899176</td>\n",
       "      <td>1.944858</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 363046 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ImmGenATAC1219.peakID  ImmGenATAC1219.peak_3  ImmGenATAC1219.peak_4  \\\n",
       "LTHSC.34-.BM                        1.748461               0.495695   \n",
       "LTHSC.34+.BM                        0.137504               0.137504   \n",
       "STHSC.150-.BM                       0.925999               0.150560   \n",
       "MPP4.135+.BM                        0.150560               0.150560   \n",
       "proB.CLP.BM                         0.555816               0.839960   \n",
       "\n",
       "ImmGenATAC1219.peakID  ImmGenATAC1219.peak_5  ImmGenATAC1219.peak_6  \\\n",
       "LTHSC.34-.BM                        0.495695               1.748461   \n",
       "LTHSC.34+.BM                        0.137504               1.400538   \n",
       "STHSC.150-.BM                       0.150560               1.510962   \n",
       "MPP4.135+.BM                        0.150560               1.367371   \n",
       "proB.CLP.BM                         0.555816               1.084064   \n",
       "\n",
       "ImmGenATAC1219.peakID  ImmGenATAC1219.peak_7  ImmGenATAC1219.peak_8  \\\n",
       "LTHSC.34-.BM                        0.495695               1.748461   \n",
       "LTHSC.34+.BM                        0.137504               0.773996   \n",
       "STHSC.150-.BM                       0.150560               0.925999   \n",
       "MPP4.135+.BM                        0.150560               0.150560   \n",
       "proB.CLP.BM                         0.150560               1.361768   \n",
       "\n",
       "ImmGenATAC1219.peakID  ImmGenATAC1219.peak_9  ImmGenATAC1219.peak_10  \\\n",
       "LTHSC.34-.BM                        2.424922                3.000000   \n",
       "LTHSC.34+.BM                        0.137504                0.137504   \n",
       "STHSC.150-.BM                       0.925999                0.150560   \n",
       "MPP4.135+.BM                        0.150560                0.150560   \n",
       "proB.CLP.BM                         0.505891                0.555816   \n",
       "\n",
       "ImmGenATAC1219.peakID  ImmGenATAC1219.peak_12  ImmGenATAC1219.peak_14  ...  \\\n",
       "LTHSC.34-.BM                         1.748461                0.495695  ...   \n",
       "LTHSC.34+.BM                         0.137504                0.137504  ...   \n",
       "STHSC.150-.BM                        0.150560                0.925999  ...   \n",
       "MPP4.135+.BM                         0.871844                0.871844  ...   \n",
       "proB.CLP.BM                          1.555816                0.555816  ...   \n",
       "\n",
       "ImmGenATAC1219.peakID  ImmGenATAC1219.peak_512584  ImmGenATAC1219.peak_512585  \\\n",
       "LTHSC.34-.BM                             1.748461                    1.748461   \n",
       "LTHSC.34+.BM                             3.072106                    1.835924   \n",
       "STHSC.150-.BM                            1.883621                    1.510962   \n",
       "MPP4.135+.BM                             2.965323                    1.655352   \n",
       "proB.CLP.BM                              4.192983                    3.768714   \n",
       "\n",
       "ImmGenATAC1219.peakID  ImmGenATAC1219.peak_512586  ImmGenATAC1219.peak_512587  \\\n",
       "LTHSC.34-.BM                             2.424922                    0.495695   \n",
       "LTHSC.34+.BM                             3.844988                    1.835924   \n",
       "STHSC.150-.BM                            3.761285                    1.883621   \n",
       "MPP4.135+.BM                             3.628774                    1.839960   \n",
       "proB.CLP.BM                              4.397803                    2.480265   \n",
       "\n",
       "ImmGenATAC1219.peakID  ImmGenATAC1219.peak_512588  ImmGenATAC1219.peak_512589  \\\n",
       "LTHSC.34-.BM                             0.495695                    0.495695   \n",
       "LTHSC.34+.BM                             3.493135                    3.072106   \n",
       "STHSC.150-.BM                            2.176323                    1.510962   \n",
       "MPP4.135+.BM                             1.655352                    2.166715   \n",
       "proB.CLP.BM                              1.959770                    2.757023   \n",
       "\n",
       "ImmGenATAC1219.peakID  ImmGenATAC1219.peak_512590  ImmGenATAC1219.peak_512591  \\\n",
       "LTHSC.34-.BM                             1.748461                    2.424922   \n",
       "LTHSC.34+.BM                             3.291309                    3.291309   \n",
       "STHSC.150-.BM                            0.925999                    2.176323   \n",
       "MPP4.135+.BM                             1.655352                    1.839960   \n",
       "proB.CLP.BM                              2.087463                    2.381283   \n",
       "\n",
       "ImmGenATAC1219.peakID  ImmGenATAC1219.peak_512592  ImmGenATAC1219.peak_512595  \n",
       "LTHSC.34-.BM                             0.495695                    0.495695  \n",
       "LTHSC.34+.BM                             3.072106                    0.773996  \n",
       "STHSC.150-.BM                            2.176323                    0.925999  \n",
       "MPP4.135+.BM                             2.014355                    2.166715  \n",
       "proB.CLP.BM                              2.899176                    1.944858  \n",
       "\n",
       "[5 rows x 363046 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ATAC2=ATAC.select_dtypes(include='number').T.drop(['Summit', 'mm10.60way.phastCons_scores', '_-log10_bestPvalue', 'Included.in.systematic.analysis'], axis=0).head(10)\n",
    "ATAC2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e4562e01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Summit                             7.593955e+07\n",
       "mm10.60way.phastCons_scores        1.773167e-01\n",
       "_-log10_bestPvalue                 6.582067e+00\n",
       "Included.in.systematic.analysis    1.000000e+00\n",
       "LTHSC.34-.BM                       1.640108e+00\n",
       "                                       ...     \n",
       "FRC.SLN                            1.724343e+00\n",
       "IAP.SLN                            1.831062e+00\n",
       "BEC.SLN                            1.710356e+00\n",
       "LEC.SLN                            1.699517e+00\n",
       "Ep.MEChi.Th                        1.772441e+00\n",
       "Length: 94, dtype: float64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ATAC.select_dtypes(include='number').mean(axis=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "19a795fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pearson_correlation_QC_metrics (plot_type, ATAC_data, QC_data, stat_type='mean', qc_metric_col=None):\n",
    "\n",
    "    '''\n",
    "    df1: first Dataframe (ATAC)\n",
    "    df2: second Dataframe (QC)\n",
    "    qc_metrics_col: welche qc metric soll korreliert werden, spaltenname\n",
    "    plot_type: scatterplot/heatmap\n",
    "    '''\n",
    "    atac_data=ATAC_data.copy()\n",
    "    \n",
    "    atac_data=atac_data.select_dtypes(include='number')\n",
    "    atac_data=atac_data.T\n",
    "    #print(atac_data.head())\n",
    "    atac_data.drop(['Summit', 'mm10.60way.phastCons_scores', '_-log10_bestPvalue', 'Included.in.systematic.analysis'], axis=0, inplace=True)\n",
    "    #print(atac_data.index)\n",
    "    #print(atac_data.columns)\n",
    "    qc_data=QC_data.copy() #damit eingabedaten nicht verändert werden\n",
    "    \n",
    "    #Index festlegen\n",
    "    \n",
    "    \n",
    "    qc_data.set_index('CellType', inplace=True)\n",
    "    # print(qc_data.index)\n",
    "    # print(qc_data.columns)\n",
    "    \n",
    "    \n",
    "    #atac_data.set_index('ImmGenATAC1219.peakID', inplace=True)\n",
    "    #atac_data.set_index(atac_data.columns[0], inplace=True)\n",
    "    # print(qc_data.index)\n",
    "\n",
    "    # print(atac_data.index)\n",
    "\n",
    "    # Einheitliche Formatierung der Zelltypen\n",
    "    # atac_data.columns = atac_data.columns.astype(str).str.lower().str.replace(\" \", \"\")\n",
    "    # qc_data.columns = qc_data.columns.astype(str).str.lower().str.replace(\" \", \"\")\n",
    "\n",
    "    #qc_data.index = qc_data.index.astype(str).str.lower().str.replace(\" \", \"\")\n",
    "\n",
    "    # Gemeinsame Zelltypen identifizieren\n",
    "    gemeinsame_zelltypen = set(atac_data.index).intersection(qc_data.index)\n",
    "    print(f\"Anzahl gemeinsamer Zelltypen: {len(gemeinsame_zelltypen)}\")\n",
    "    print(\"Gemeinsame Zelltypen (Beispiel):\", list(gemeinsame_zelltypen)[:10])\n",
    "\n",
    "    # Zelltypen, die nur in einem der DataFrames vorkommen\n",
    "    print(\"\\nZelltypen nur in ATAC:\", list(set(atac_data.index) - set(qc_data.index))[:10])\n",
    "    print(\"Zelltypen nur in QC:\", list(set(qc_data.index) - set(atac_data.index))[:10])\n",
    "\n",
    "    # QC = QC.drop(QC.columns[0:2], axis=1)\n",
    "    # QC.index = QC.iloc[:, 1].str.lower().str.replace(\" \", \"\")\n",
    "    #print(stat_type)\n",
    "    #return atac_data, qc_data\n",
    "    if stat_type=='mean':\n",
    "        atac_stat=atac_data.mean(axis=1)#.rename('ATAC_stat')\n",
    "    elif stat_type=='median':\n",
    "        atac_stat=atac_data.median(axis=1)#.rename('ATAC_stat')\n",
    "    elif stat_type=='std':\n",
    "        atac_stat=atac_data.std(axis=1)#.rename('ATAC_stat')\n",
    "\n",
    "    print(atac_stat)\n",
    "    print(qc_data.head())\n",
    "\n",
    "    # if qc_metric_col not in qc_data.columns:\n",
    "    #     raise ValueError(f\"QC-Metrik '{qc_metric_col}' nicht gefunden.\")\n",
    "    \n",
    "    # qc_metric = qc_data[qc_metric_col]\n",
    "    # print(qc_metric)\n",
    "    # df=pd.concat([atac_stat, qc_metric], axis=1, join='inner')\n",
    "    # atac_stat['qc_metric']=qc_data[qc_metric_col]\n",
    "    # df.columns=['ATAC_stat', qc_metric_col]\n",
    "\n",
    "    # if len(df)<2:\n",
    "    #    raise ValueError(\"Nicht genügend überlappende Zelltypen für Korrelation\")\n",
    "\n",
    "\n",
    "    # # qc_metric=qc_data.set_index('CellType') [[qc_metric_col]]\n",
    "    # # merged=pd.merge(atac_means, qc_metric, left_index=True, right_index=True)\n",
    "    # r, p=pearsonr(df['ATAC_stat'], df[qc_metric_col])\n",
    "    # print(f\"pearson-correlation ({stat_type} vs {qc_metric_col}): r= {r:.4f}, p-Wert: {p:.4e}\")\n",
    "     \n",
    "    #         #Scatterplot\n",
    "    # if plot_type.lower()=='scatterplot':\n",
    "    #     plt.figure(figsize=(6,4))\n",
    "    #     sns.scatterplot(data=df, x='ATAC_mean', y=qc_metric_col)\n",
    "    #     slope, intercept, *_ = linregress(df['ATAC_mean'], df[qc_metric_col])\n",
    "    #     x_vals = pd.Series(sorted(df['ATAC_mean']))\n",
    "    #     y_vals = slope * x_vals + intercept\n",
    "    #     plt.plot(x_vals, y_vals, color='purple', label='Regression')\n",
    "    #     plt.xlabel('ATAC seq mean')\n",
    "    #     plt.ylabel(qc_metric_col)\n",
    "    #     plt.title(f'ATAC seq-{qc_metric_col}')\n",
    "    #     plt.legend()\n",
    "    #     plt.tight_layout()\n",
    "    #     plt.show()\n",
    "            \n",
    "    #         #heatmap\n",
    "    # elif plot_type.lower()=='heatmap':\n",
    "    #     plt.figure(figsize=(3,3))\n",
    "    #     sns.heatmap([[r]], annot=True, cmap='coolwarm', cbar=True, xticklabels=[qc_metric_col], yticklabels=[\"ATAC_mean\"])\n",
    "    #     plt.title('Pearson correlation')\n",
    "    #     plt.tight_layout()\n",
    "    #     plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0324259",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(90, 363046)\n",
      "(181, 12)\n",
      "Anzahl gemeinsamer Zelltypen: 0\n",
      "Gemeinsame Zelltypen (Beispiel): []\n",
      "\n",
      "Zelltypen nur in ATAC: ['LTHSC.34+.BM', 'T8.TE.LCMV.d7.Sp', 'preT.DN3.Th', 'proB.FrA.BM', 'T8.IEL.LCMV.d7.SI', 'Mo.6C+II-.Bl', 'proB.FrBC.BM', 'DC.8+.Sp', 'MF.microglia.CNS', 'B.PB.Sp']\n",
      "Zelltypen nur in QC: ['IAP.SLN#2', 'Treg.4.FP3+.Nrplo.Co#1', 'DC.103+11b-.SI#2', 'T.4.Th#1', 'proB.CLP.BM#2', 'GN.Thio.PC#3', 'T8.MP.LCMV.d7.Sp#1', 'B1b.PC#2', 'T.4.Nve.Fem.Sp#1', 'ILC3.CCR6+.SI#1']\n",
      "(0, 363046)\n",
      "(0,)\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "`x` and `y` must have length at least 2.",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mValueError\u001b[39m                                Traceback (most recent call last)",
      "\u001b[32m~\\AppData\\Local\\Temp\\ipykernel_10896\\2265129630.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m()\u001b[39m\n\u001b[32m     57\u001b[39m \u001b[38;5;66;03m# print(\"QC-Werte Länge:\", len(qc_vals))\u001b[39;00m\n\u001b[32m     58\u001b[39m \u001b[38;5;66;03m# print(\"Fehlende Werte in QC:\", qc_vals.isna().sum())\u001b[39;00m\n\u001b[32m     59\u001b[39m \u001b[38;5;66;03m# print(\"Fehlende Werte in ATAC:\", atac_vals.isna().sum())\u001b[39;00m\n\u001b[32m     60\u001b[39m \n\u001b[32m---> \u001b[39m\u001b[32m61\u001b[39m r, p_value=pearsonr(atac_vals, qc_vals)\n\u001b[32m     62\u001b[39m print(f\"Pearson r = {r:.3f}, p = {p_value:.3g}\")\n",
      "\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\scipy\\stats\\_stats_py.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m(x, y, alternative, method, axis)\u001b[39m\n\u001b[32m   4548\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m n != y.shape[axis]:\n\u001b[32m   4549\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m ValueError(\u001b[33m'`x` and `y` must have the same length along `axis`.'\u001b[39m)\n\u001b[32m   4550\u001b[39m \n\u001b[32m   4551\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m n < \u001b[32m2\u001b[39m:\n\u001b[32m-> \u001b[39m\u001b[32m4552\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m ValueError(\u001b[33m'`x` and `y` must have length at least 2.'\u001b[39m)\n\u001b[32m   4553\u001b[39m \n\u001b[32m   4554\u001b[39m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m   4555\u001b[39m         x, y = xp.broadcast_arrays(x, y)\n",
      "\u001b[31mValueError\u001b[39m: `x` and `y` must have length at least 2."
     ]
    }
   ],
   "source": [
    "from scipy.stats import pearsonr\n",
    "atac_data=ATAC.copy()\n",
    "\n",
    "    \n",
    "atac_data=atac_data.select_dtypes(include='number')\n",
    "atac_data=atac_data.T\n",
    "#print(atac_data.head())\n",
    "atac_data.drop(['Summit', 'mm10.60way.phastCons_scores', '_-log10_bestPvalue', 'Included.in.systematic.analysis'], axis=0)\n",
    "#print(atac_data.index)\n",
    "#print(atac_data.columns)\n",
    "qc_data=QC.copy() #damit eingabedaten nicht verändert werden\n",
    "\n",
    "#Index festlegen\n",
    "\n",
    "\n",
    "qc_data.set_index('CellType')\n",
    "# print(qc_data.index)\n",
    "# print(qc_data.columns)\n",
    "\n",
    "print(atac_data.shape)\n",
    "print(qc_data.shape)\n",
    "\n",
    "#atac_data.set_index('ImmGenATAC1219.peakID', inplace=True)\n",
    "#atac_data.set_index(atac_data.columns[0], inplace=True)\n",
    "# print(qc_data.index)\n",
    "\n",
    "# print(atac_data.index)\n",
    "\n",
    "# Einheitliche Formatierung der Zelltypen\n",
    "# atac_data.columns = atac_data.columns.astype(str).str.lower().str.replace(\" \", \"\")\n",
    "# qc_data.columns = qc_data.columns.astype(str).str.lower().str.replace(\" \", \"\")\n",
    "\n",
    "#qc_data.index = qc_data.index.astype(str).str.lower().str.replace(\" \", \"\")\n",
    "\n",
    "# Gemeinsame Zelltypen identifizieren\n",
    "#gemeinsame_zelltypen = set(atac_data.index).intersection(qc_data.index)\n",
    "print(f\"Anzahl gemeinsamer Zelltypen: {len(gemeinsame_zelltypen)}\")\n",
    "print(\"Gemeinsame Zelltypen (Beispiel):\", list(gemeinsame_zelltypen)[:10])\n",
    "\n",
    "# Zelltypen, die nur in einem der DataFrames vorkommen\n",
    "print(\"\\nZelltypen nur in ATAC:\", list(set(atac_data.index) - set(qc_data.index))[:10])\n",
    "print(\"Zelltypen nur in QC:\", list(set(qc_data.index) - set(atac_data.index))[:10])\n",
    "\n",
    "atac_stat=atac_data.mean(axis=1)#.rename('ATAC_stat')\n",
    "\n",
    "gemeinsame_zelltypen = list((atac_data.index).intersection(qc_data.index))\n",
    "\n",
    "#common_cells = atac_stat.index.intersection(qc_data.index)\n",
    "atac_vals = atac_data.loc[gemeinsame_zelltypen]\n",
    "#print(atac_vals.head(10))\n",
    "print(atac_vals.shape)\n",
    "qc_metric_col = 'InputCellNumber'\n",
    "qc_vals = qc_data.loc[gemeinsame_zelltypen, qc_metric_col]\n",
    "print(qc_vals.shape)\n",
    "#print(qc_vals.head(10))\n",
    "# print(\"ATAC-Werte Länge:\", len(atac_vals))\n",
    "# print(\"QC-Werte Länge:\", len(qc_vals))\n",
    "# print(\"Fehlende Werte in QC:\", qc_vals.isna().sum())\n",
    "# print(\"Fehlende Werte in ATAC:\", atac_vals.isna().sum())\n",
    "\n",
    "r, p_value=pearsonr(atac_vals, qc_vals)\n",
    "print(f\"Pearson r = {r:.3f}, p = {p_value:.3g}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d2d186c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Anzahl gemeinsamer Zelltypen: 89\n",
      "Gemeinsame Zelltypen (Beispiel): ['LTHSC.34+.BM', 'T8.TE.LCMV.d7.Sp', 'preT.DN3.Th', 'proB.FrA.BM', 'T8.IEL.LCMV.d7.SI', 'Mo.6C+II-.Bl', 'proB.FrBC.BM', 'DC.8+.Sp', 'MF.microglia.CNS', 'B.PB.Sp']\n",
      "\n",
      "Zelltypen nur in ATAC: ['FRC.SLN']\n",
      "Zelltypen nur in QC: ['FRC.CD140a+.Madcam-.CD35-.SLN']\n",
      "LTHSC.34-.BM     1.209332\n",
      "LTHSC.34+.BM     1.258308\n",
      "STHSC.150-.BM    1.263741\n",
      "MPP4.135+.BM     1.277657\n",
      "proB.CLP.BM      1.302637\n",
      "                   ...   \n",
      "FRC.SLN          1.258732\n",
      "IAP.SLN          1.318027\n",
      "BEC.SLN          1.258927\n",
      "LEC.SLN          1.261929\n",
      "Ep.MEChi.Th      1.278902\n",
      "Length: 90, dtype: float64\n",
      "              ImmGenLab    Lineage      CellFamily        Organ  \\\n",
      "CellType                                                          \n",
      "LTHSC.34-.BM     Wagers  Stem&Prog  Stem&Prog Cell  Bone Marrow   \n",
      "LTHSC.34+.BM     Wagers  Stem&Prog  Stem&Prog Cell  Bone Marrow   \n",
      "STHSC.150-.BM    Wagers  Stem&Prog  Stem&Prog Cell  Bone Marrow   \n",
      "MPP4.135+.BM     Wagers  Stem&Prog  Stem&Prog Cell  Bone Marrow   \n",
      "proB.CLP.BM       Hardy          B          B Cell  Bone Marrow   \n",
      "\n",
      "                                                  SortingMarkers  \\\n",
      "CellType                                                           \n",
      "LTHSC.34-.BM                Lin-Sca1+ckit+CD135-CD150+CD48-CD34-   \n",
      "LTHSC.34+.BM                Lin-Sca1+ckit+CD135-CD150+CD48-CD34+   \n",
      "STHSC.150-.BM                    Lin-Sca1+ckit+CD135-CD150-CD48-   \n",
      "MPP4.135+.BM                                Lin-Sca1+ckit+CD135+   \n",
      "proB.CLP.BM    CD19-IgM-CD43+CD24-CD45R-CD93+CD117+IL7Ra+PI-D...   \n",
      "\n",
      "               InputCellNumber  PF.reads  %chrM.mapped  \\\n",
      "CellType                                                 \n",
      "LTHSC.34-.BM               677  22287984          4.58   \n",
      "LTHSC.34+.BM              2483  28588536          3.17   \n",
      "STHSC.150-.BM             3660  22170346          3.25   \n",
      "MPP4.135+.BM             10000  24003590         13.83   \n",
      "proB.CLP.BM              10000  23578208          7.67   \n",
      "\n",
      "               Paired.read.after.removing.PCR.duplication  %fragment.1Kb_TSS  \\\n",
      "CellType                                                                       \n",
      "LTHSC.34-.BM                                      2470102              10.60   \n",
      "LTHSC.34+.BM                                      7675501               7.70   \n",
      "STHSC.150-.BM                                     6060045              11.13   \n",
      "MPP4.135+.BM                                      6128252              18.96   \n",
      "proB.CLP.BM                                       6588564              13.21   \n",
      "\n",
      "               Replicate.cor  \n",
      "CellType                      \n",
      "LTHSC.34-.BM             NaN  \n",
      "LTHSC.34+.BM             NaN  \n",
      "STHSC.150-.BM            NaN  \n",
      "MPP4.135+.BM             NaN  \n",
      "proB.CLP.BM            0.868  \n"
     ]
    }
   ],
   "source": [
    "pearson_correlation_QC_metrics (plot_type='heatmap', ATAC_data=ATAC, QC_data=QC, stat_type='mean', qc_metric_col='InputCellNumber')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "d7ed251d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CellType\n",
       "LTHSC.34-.BM       677\n",
       "LTHSC.34+.BM      2483\n",
       "STHSC.150-.BM     3660\n",
       "MPP4.135+.BM     10000\n",
       "proB.CLP.BM      10000\n",
       "                 ...  \n",
       "BEC.SLN          10000\n",
       "LEC.SLN          10000\n",
       "LEC.SLN          10000\n",
       "Ep.MEChi.Th      10000\n",
       "Ep.MEChi.Th      10000\n",
       "Name: InputCellNumber, Length: 181, dtype: int64"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qc_data['InputCellNumber']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "dc7692cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Anzahl gemeinsamer Zelltypen: 89\n",
      "Gemeinsame Zelltypen (Beispiel): ['T.DN4.Th', 'B.FrE.BM', 'MPP4.135+.BM', 'DC.pDC.Sp', 'GN.BM', 'B.PB.Sp', 'MPP3.48+.BM', 'T.DP.Th', 'preB.FrD.BM', 'preT.DN2a.Th']\n",
      "\n",
      "Zelltypen nur in ATAC: ['FRC.SLN']\n",
      "Zelltypen nur in QC: ['FRC.CD140a+.Madcam-.CD35-.SLN']\n",
      "LTHSC.34-.BM     1.640108\n",
      "LTHSC.34+.BM     1.739950\n",
      "STHSC.150-.BM    1.751400\n",
      "MPP4.135+.BM     1.780920\n",
      "proB.CLP.BM      1.823342\n",
      "                   ...   \n",
      "FRC.SLN          1.724343\n",
      "IAP.SLN          1.831062\n",
      "BEC.SLN          1.710356\n",
      "LEC.SLN          1.699517\n",
      "Ep.MEChi.Th      1.772441\n",
      "Length: 90, dtype: float64\n",
      "CellType\n",
      "LTHSC.34-.BM       677\n",
      "LTHSC.34+.BM      2483\n",
      "STHSC.150-.BM     3660\n",
      "MPP4.135+.BM     10000\n",
      "proB.CLP.BM      10000\n",
      "                 ...  \n",
      "BEC.SLN          10000\n",
      "LEC.SLN          10000\n",
      "LEC.SLN          10000\n",
      "Ep.MEChi.Th      10000\n",
      "Ep.MEChi.Th      10000\n",
      "Name: InputCellNumber, Length: 181, dtype: int64\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "cannot unpack non-iterable NoneType object",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mTypeError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[80]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m atac_data, qc_data=pearson_correlation_QC_metrics (plot_type=\u001b[33m'\u001b[39m\u001b[33mheatmap\u001b[39m\u001b[33m'\u001b[39m, ATAC_data=ATAC, QC_data=QC, stat_type=\u001b[33m'\u001b[39m\u001b[33mmean\u001b[39m\u001b[33m'\u001b[39m, qc_metric_col=\u001b[33m'\u001b[39m\u001b[33mInputCellNumber\u001b[39m\u001b[33m'\u001b[39m)\n",
      "\u001b[31mTypeError\u001b[39m: cannot unpack non-iterable NoneType object"
     ]
    }
   ],
   "source": [
    "atac_data, qc_data=pearson_correlation_QC_metrics (plot_type='heatmap', ATAC_data=ATAC, QC_data=QC, stat_type='mean', qc_metric_col='InputCellNumber')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "4aee784a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "LTHSC.34-.BM     1.640108\n",
       "LTHSC.34+.BM     1.739950\n",
       "STHSC.150-.BM    1.751400\n",
       "MPP4.135+.BM     1.780920\n",
       "proB.CLP.BM      1.823342\n",
       "                   ...   \n",
       "FRC.SLN          1.724343\n",
       "IAP.SLN          1.831062\n",
       "BEC.SLN          1.710356\n",
       "LEC.SLN          1.699517\n",
       "Ep.MEChi.Th      1.772441\n",
       "Length: 90, dtype: float64"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#atac_data.columns\n",
    "atac_data.mean(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cccc67d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CellType</th>\n",
       "      <th>ImmGenLab</th>\n",
       "      <th>Lineage</th>\n",
       "      <th>CellFamily</th>\n",
       "      <th>Organ</th>\n",
       "      <th>SortingMarkers</th>\n",
       "      <th>InputCellNumber</th>\n",
       "      <th>PF.reads</th>\n",
       "      <th>%chrM.mapped</th>\n",
       "      <th>Paired.read.after.removing.PCR.duplication</th>\n",
       "      <th>%fragment.1Kb_TSS</th>\n",
       "      <th>Replicate.cor</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SampleName</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LTHSC.34-.BM#1</th>\n",
       "      <td>LTHSC.34-.BM</td>\n",
       "      <td>Wagers</td>\n",
       "      <td>Stem&amp;Prog</td>\n",
       "      <td>Stem&amp;Prog Cell</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>Lin-Sca1+ckit+CD135-CD150+CD48-CD34-</td>\n",
       "      <td>677</td>\n",
       "      <td>22287984</td>\n",
       "      <td>4.58</td>\n",
       "      <td>2470102</td>\n",
       "      <td>10.60</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LTHSC.34+.BM#1</th>\n",
       "      <td>LTHSC.34+.BM</td>\n",
       "      <td>Wagers</td>\n",
       "      <td>Stem&amp;Prog</td>\n",
       "      <td>Stem&amp;Prog Cell</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>Lin-Sca1+ckit+CD135-CD150+CD48-CD34+</td>\n",
       "      <td>2483</td>\n",
       "      <td>28588536</td>\n",
       "      <td>3.17</td>\n",
       "      <td>7675501</td>\n",
       "      <td>7.70</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STHSC.150-.BM#1</th>\n",
       "      <td>STHSC.150-.BM</td>\n",
       "      <td>Wagers</td>\n",
       "      <td>Stem&amp;Prog</td>\n",
       "      <td>Stem&amp;Prog Cell</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>Lin-Sca1+ckit+CD135-CD150-CD48-</td>\n",
       "      <td>3660</td>\n",
       "      <td>22170346</td>\n",
       "      <td>3.25</td>\n",
       "      <td>6060045</td>\n",
       "      <td>11.13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MPP4.135+.BM#1</th>\n",
       "      <td>MPP4.135+.BM</td>\n",
       "      <td>Wagers</td>\n",
       "      <td>Stem&amp;Prog</td>\n",
       "      <td>Stem&amp;Prog Cell</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>Lin-Sca1+ckit+CD135+</td>\n",
       "      <td>10000</td>\n",
       "      <td>24003590</td>\n",
       "      <td>13.83</td>\n",
       "      <td>6128252</td>\n",
       "      <td>18.96</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>proB.CLP.BM#1</th>\n",
       "      <td>proB.CLP.BM</td>\n",
       "      <td>Hardy</td>\n",
       "      <td>B</td>\n",
       "      <td>B Cell</td>\n",
       "      <td>Bone Marrow</td>\n",
       "      <td>CD19-IgM-CD43+CD24-CD45R-CD93+CD117+IL7Ra+PI-D...</td>\n",
       "      <td>10000</td>\n",
       "      <td>23578208</td>\n",
       "      <td>7.67</td>\n",
       "      <td>6588564</td>\n",
       "      <td>13.21</td>\n",
       "      <td>0.868</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      CellType ImmGenLab    Lineage      CellFamily  \\\n",
       "SampleName                                                            \n",
       "LTHSC.34-.BM#1    LTHSC.34-.BM    Wagers  Stem&Prog  Stem&Prog Cell   \n",
       "LTHSC.34+.BM#1    LTHSC.34+.BM    Wagers  Stem&Prog  Stem&Prog Cell   \n",
       "STHSC.150-.BM#1  STHSC.150-.BM    Wagers  Stem&Prog  Stem&Prog Cell   \n",
       "MPP4.135+.BM#1    MPP4.135+.BM    Wagers  Stem&Prog  Stem&Prog Cell   \n",
       "proB.CLP.BM#1      proB.CLP.BM     Hardy          B          B Cell   \n",
       "\n",
       "                       Organ  \\\n",
       "SampleName                     \n",
       "LTHSC.34-.BM#1   Bone Marrow   \n",
       "LTHSC.34+.BM#1   Bone Marrow   \n",
       "STHSC.150-.BM#1  Bone Marrow   \n",
       "MPP4.135+.BM#1   Bone Marrow   \n",
       "proB.CLP.BM#1    Bone Marrow   \n",
       "\n",
       "                                                    SortingMarkers  \\\n",
       "SampleName                                                           \n",
       "LTHSC.34-.BM#1                Lin-Sca1+ckit+CD135-CD150+CD48-CD34-   \n",
       "LTHSC.34+.BM#1                Lin-Sca1+ckit+CD135-CD150+CD48-CD34+   \n",
       "STHSC.150-.BM#1                    Lin-Sca1+ckit+CD135-CD150-CD48-   \n",
       "MPP4.135+.BM#1                                Lin-Sca1+ckit+CD135+   \n",
       "proB.CLP.BM#1    CD19-IgM-CD43+CD24-CD45R-CD93+CD117+IL7Ra+PI-D...   \n",
       "\n",
       "                 InputCellNumber  PF.reads  %chrM.mapped  \\\n",
       "SampleName                                                 \n",
       "LTHSC.34-.BM#1               677  22287984          4.58   \n",
       "LTHSC.34+.BM#1              2483  28588536          3.17   \n",
       "STHSC.150-.BM#1             3660  22170346          3.25   \n",
       "MPP4.135+.BM#1             10000  24003590         13.83   \n",
       "proB.CLP.BM#1              10000  23578208          7.67   \n",
       "\n",
       "                 Paired.read.after.removing.PCR.duplication  \\\n",
       "SampleName                                                    \n",
       "LTHSC.34-.BM#1                                      2470102   \n",
       "LTHSC.34+.BM#1                                      7675501   \n",
       "STHSC.150-.BM#1                                     6060045   \n",
       "MPP4.135+.BM#1                                      6128252   \n",
       "proB.CLP.BM#1                                       6588564   \n",
       "\n",
       "                 %fragment.1Kb_TSS  Replicate.cor  \n",
       "SampleName                                         \n",
       "LTHSC.34-.BM#1               10.60            NaN  \n",
       "LTHSC.34+.BM#1                7.70            NaN  \n",
       "STHSC.150-.BM#1              11.13            NaN  \n",
       "MPP4.135+.BM#1               18.96            NaN  \n",
       "proB.CLP.BM#1                13.21          0.868  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "QC = data['QC_metrics']\n",
    "QC.head()\n",
    "# qc_data=QC\n",
    "# qc_data=qc_data.iloc[:, 1:]\n",
    "# qc_data.index=qc_data.iloc[:, 0:]\n",
    "# qc_data.index = qc_data.index.astype(str).str.lower().str.replace(\" \", \"\")\n",
    "# #qc_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f72143d",
   "metadata": {},
   "source": [
    "was muss ich in pearson r reinstecken: von hand machen!\n",
    "größe reduzieren\n",
    "\n",
    "ist functions falsch?\n",
    "\n",
    "zeilenweise aus functions rauskopieren udn immer raus geben\n",
    "\n",
    "return atac data\n",
    "\n",
    "von hand die function nicht als funktion, von hand basteln -> wo fehler\n",
    "\n",
    "von hand überprüfen ob das richtige ruaskommt\n",
    "\n",
    "zb gucken ob das mit erste spalte löschen geklappt hat, schritt für schritt\n",
    "\n",
    "erst von hand dann function\n",
    "\n",
    "clean up in functions aus meiner funktion löschen, damit es allgemeiner ist\n",
    "sonst meine funktion zweiteilen und eigenes cleanup machen(oberen teil)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b88aa26",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1=ATAC.head(20)\n",
    "df2=QC.head(20)\n",
    "\n",
    "df1.mean(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "79f012cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ImmGenATAC1219.peakID           ImmGenATAC1219.peak_3 ImmGenATAC1219.peak_4  \\\n",
      "chrom                                            chr1                  chr1   \n",
      "Summit                                        3120109               3121485   \n",
      "mm10.60way.phastCons_scores                      0.07                  0.15   \n",
      "_-log10_bestPvalue                               10.8                  3.02   \n",
      "Included.in.systematic.analysis                   1.0                   1.0   \n",
      "\n",
      "ImmGenATAC1219.peakID           ImmGenATAC1219.peak_5 ImmGenATAC1219.peak_6  \\\n",
      "chrom                                            chr1                  chr1   \n",
      "Summit                                        3372787               3399217   \n",
      "mm10.60way.phastCons_scores                      0.03                  0.06   \n",
      "_-log10_bestPvalue                               1.31                  2.39   \n",
      "Included.in.systematic.analysis                   NaN                   1.0   \n",
      "\n",
      "ImmGenATAC1219.peakID           ImmGenATAC1219.peak_7 ImmGenATAC1219.peak_8  \\\n",
      "chrom                                            chr1                  chr1   \n",
      "Summit                                        3400115               3416260   \n",
      "mm10.60way.phastCons_scores                      0.44                  0.01   \n",
      "_-log10_bestPvalue                               2.57                  2.57   \n",
      "Included.in.systematic.analysis                   1.0                   1.0   \n",
      "\n",
      "ImmGenATAC1219.peakID           ImmGenATAC1219.peak_9 ImmGenATAC1219.peak_10  \\\n",
      "chrom                                            chr1                   chr1   \n",
      "Summit                                        3434092                3434378   \n",
      "mm10.60way.phastCons_scores                      0.18                   0.15   \n",
      "_-log10_bestPvalue                               1.34                   1.83   \n",
      "Included.in.systematic.analysis                   NaN                    1.0   \n",
      "\n",
      "ImmGenATAC1219.peakID           ImmGenATAC1219.peak_12 ImmGenATAC1219.peak_14  \\\n",
      "chrom                                             chr1                   chr1   \n",
      "Summit                                         3446022                3514761   \n",
      "mm10.60way.phastCons_scores                        0.0                   0.07   \n",
      "_-log10_bestPvalue                                2.23                   1.83   \n",
      "Included.in.systematic.analysis                    1.0                    1.0   \n",
      "\n",
      "ImmGenATAC1219.peakID            ... ImmGenATAC1219.peak_512584  \\\n",
      "chrom                            ...                       chrY   \n",
      "Summit                           ...                   90807755   \n",
      "mm10.60way.phastCons_scores      ...                        0.0   \n",
      "_-log10_bestPvalue               ...                       9.91   \n",
      "Included.in.systematic.analysis  ...                        1.0   \n",
      "\n",
      "ImmGenATAC1219.peakID           ImmGenATAC1219.peak_512585  \\\n",
      "chrom                                                 chrY   \n",
      "Summit                                            90808523   \n",
      "mm10.60way.phastCons_scores                            0.0   \n",
      "_-log10_bestPvalue                                   19.71   \n",
      "Included.in.systematic.analysis                        1.0   \n",
      "\n",
      "ImmGenATAC1219.peakID           ImmGenATAC1219.peak_512586  \\\n",
      "chrom                                                 chrY   \n",
      "Summit                                            90808861   \n",
      "mm10.60way.phastCons_scores                            0.0   \n",
      "_-log10_bestPvalue                                    6.58   \n",
      "Included.in.systematic.analysis                        1.0   \n",
      "\n",
      "ImmGenATAC1219.peakID           ImmGenATAC1219.peak_512587  \\\n",
      "chrom                                                 chrY   \n",
      "Summit                                            90811022   \n",
      "mm10.60way.phastCons_scores                           0.04   \n",
      "_-log10_bestPvalue                                    4.35   \n",
      "Included.in.systematic.analysis                        1.0   \n",
      "\n",
      "ImmGenATAC1219.peakID           ImmGenATAC1219.peak_512588  \\\n",
      "chrom                                                 chrY   \n",
      "Summit                                            90811406   \n",
      "mm10.60way.phastCons_scores                            0.0   \n",
      "_-log10_bestPvalue                                     2.6   \n",
      "Included.in.systematic.analysis                        1.0   \n",
      "\n",
      "ImmGenATAC1219.peakID           ImmGenATAC1219.peak_512589  \\\n",
      "chrom                                                 chrY   \n",
      "Summit                                            90811728   \n",
      "mm10.60way.phastCons_scores                            0.0   \n",
      "_-log10_bestPvalue                                    2.33   \n",
      "Included.in.systematic.analysis                        1.0   \n",
      "\n",
      "ImmGenATAC1219.peakID           ImmGenATAC1219.peak_512590  \\\n",
      "chrom                                                 chrY   \n",
      "Summit                                            90812084   \n",
      "mm10.60way.phastCons_scores                            0.0   \n",
      "_-log10_bestPvalue                                    3.12   \n",
      "Included.in.systematic.analysis                        1.0   \n",
      "\n",
      "ImmGenATAC1219.peakID           ImmGenATAC1219.peak_512591  \\\n",
      "chrom                                                 chrY   \n",
      "Summit                                            90812450   \n",
      "mm10.60way.phastCons_scores                            0.0   \n",
      "_-log10_bestPvalue                                    3.99   \n",
      "Included.in.systematic.analysis                        1.0   \n",
      "\n",
      "ImmGenATAC1219.peakID           ImmGenATAC1219.peak_512592  \\\n",
      "chrom                                                 chrY   \n",
      "Summit                                            90812906   \n",
      "mm10.60way.phastCons_scores                            0.0   \n",
      "_-log10_bestPvalue                                    3.21   \n",
      "Included.in.systematic.analysis                        1.0   \n",
      "\n",
      "ImmGenATAC1219.peakID           ImmGenATAC1219.peak_512595  \n",
      "chrom                                                 chrY  \n",
      "Summit                                            90828985  \n",
      "mm10.60way.phastCons_scores                            0.0  \n",
      "_-log10_bestPvalue                                    1.41  \n",
      "Included.in.systematic.analysis                        NaN  \n",
      "\n",
      "[5 rows x 363046 columns]\n",
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: [chrom, TSS, genes.within.100Kb, LTHSC.34-.BM, LTHSC.34+.BM]\n",
      "Index(['CellType', 'ImmGenLab', 'Lineage', 'CellFamily', 'Organ',\n",
      "       'SortingMarkers', 'InputCellNumber', 'PF.reads', '%chrM.mapped',\n",
      "       'Paired.read.after.removing.PCR.duplication', '%fragment.1Kb_TSS',\n",
      "       'Replicate.cor'],\n",
      "      dtype='object')\n",
      "Index([], dtype='object', name='ImmGenATAC1219.peakID')\n",
      "Anzahl gemeinsamer Zelltypen: 0\n",
      "Gemeinsame Zelltypen (Beispiel): []\n",
      "\n",
      "Zelltypen nur in ATAC: ['T.DN4.Th', 'B.FrE.BM', 'MPP4.135+.BM', 'DC.pDC.Sp', 'GN.BM', 'B.PB.Sp', 'MPP3.48+.BM', 'T.DP.Th', 'preB.FrD.BM', 'preT.DN2a.Th']\n",
      "Zelltypen nur in QC: ['B.PC.BM#2', 'Tgd.g1.1+d1.24a+.Th#2', 'MPP4.135+.BM#1', 'B.mem.Sp#2', 'Ep.MEChi.Th#2', 'T.ISP.Th#2', 'NK.27+11b+.BM#1', 'B.Fo.Sp#2', 'Tgd.g1.1+d1.24a+.Th#1', 'MF.Fem.PC#1']\n"
     ]
    }
   ],
   "source": [
    "#plot_type, atac_data, qc_data, qc_metric_col=None\n",
    "pearson_correlation_QC_metrics(plot_type='scatterplot', ATAC_data=ATAC, QC_data=QC, stat_type='mean', qc_metric_col='InputCellNumber')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "5b349010",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from scipy.stats import spearmanr\n",
    "\n",
    "def correlate_atac_with_qc(ATAC, QC):\n",
    "    # 1. Zelltypen aus QC extrahieren:\n",
    "    # Erste Zeile in QC enthält zelltyp+`#`, zweite Zeile die echten zelltyp-Namen\n",
    "    # Also nehmen wir die zweite Zeile als neue Spaltennamen und löschen die erste Zeile\n",
    "\n",
    "    # QC Spaltennamen (erstes Header-Level)\n",
    "    qc_first_row = QC.iloc[0]\n",
    "    # Echte Zelltypen (zweite Zeile)\n",
    "    qc_second_row = QC.iloc[1]\n",
    "\n",
    "    # Ersetze Spaltennamen in QC durch zweite Zeile\n",
    "    QC.columns = qc_second_row\n",
    "    # Entferne die ersten zwei Zeilen (Header + extra Zeile)\n",
    "    QC = QC.iloc[2:].reset_index(drop=True)\n",
    "\n",
    "    # 2. Mittelwerte aus ATAC berechnen (pro Zelltyp = Spalte)\n",
    "    # Angenommen, alle Spalten in ATAC sind Zelltypen, oder filtere entsprechend:\n",
    "    atac_numeric = ATAC.select_dtypes(include='number')\n",
    "    mean_per_celltype = atac_numeric.mean(axis=0)\n",
    "\n",
    "    mean_df = mean_per_celltype.reset_index()\n",
    "    mean_df.columns = ['CellType', 'MeanValue']\n",
    "\n",
    "    # 3. QC numerische Werte auswählen (z.B. nur die numerischen Spalten)\n",
    "    qc_numeric = QC.select_dtypes(include='number')\n",
    "\n",
    "    # 4. QC Zelltypen und Mean Zelltypen auf gemeinsamen Schnitt bringen\n",
    "    # Wichtig: QC Spalten = Zelltypen, aber QC ist transponiert, daher brauchen wir Zelltypen als Spalten oder Zeilen?\n",
    "    # Nehmen wir an QC Zelltypen sind Spalten (nach Ersetzen der Spaltennamen), dann müssen wir QC transponieren, um Zelltypen als Zeilen zu haben:\n",
    "    qc_numeric = qc_numeric.transpose()\n",
    "    qc_numeric.columns = QC.index  # zelltyp-spezifische Messungen pro Metrik\n",
    "    \n",
    "    # qc_numeric index sind jetzt Zelltypen\n",
    "    qc_numeric.index.name = 'CellType'\n",
    "    qc_numeric.reset_index(inplace=True)\n",
    "\n",
    "    # 5. Zusammenführen nach Zelltyp (mean_df und qc_numeric)\n",
    "    merged = pd.merge(mean_df, qc_numeric, on='CellType', how='inner')\n",
    "\n",
    "    # 6. Spearman-Korrelation zwischen MeanValue und allen QC-Metriken berechnen\n",
    "    correlations = {}\n",
    "    for qc_metric in merged.columns:\n",
    "        if qc_metric in ['CellType', 'MeanValue']:\n",
    "            continue\n",
    "        corr, pval = spearmanr(merged['MeanValue'], merged[qc_metric])\n",
    "        correlations[qc_metric] = {'Spearman_r': corr, 'p_value': pval}\n",
    "\n",
    "    return pd.DataFrame.from_dict(correlations, orient='index')\n",
    "\n",
    "# Beispiel:\n",
    "# result = correlate_atac_with_qc(ATAC, QC)\n",
    "# print(result)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "18cf9a43",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "You are trying to merge on object and float64 columns for key 'CellType'. If you wish to proceed you should use pd.concat",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mValueError\u001b[39m                                Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[33]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43mcorrelate_atac_with_qc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mATAC\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mQC\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[32]\u001b[39m\u001b[32m, line 41\u001b[39m, in \u001b[36mcorrelate_atac_with_qc\u001b[39m\u001b[34m(ATAC, QC)\u001b[39m\n\u001b[32m     38\u001b[39m qc_numeric.reset_index(inplace=\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[32m     40\u001b[39m \u001b[38;5;66;03m# 5. Zusammenführen nach Zelltyp (mean_df und qc_numeric)\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m41\u001b[39m merged = \u001b[43mpd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mmerge\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmean_df\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mqc_numeric\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mon\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mCellType\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mhow\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43minner\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m     43\u001b[39m \u001b[38;5;66;03m# 6. Spearman-Korrelation zwischen MeanValue und allen QC-Metriken berechnen\u001b[39;00m\n\u001b[32m     44\u001b[39m correlations = {}\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\pandas\\core\\reshape\\merge.py:170\u001b[39m, in \u001b[36mmerge\u001b[39m\u001b[34m(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)\u001b[39m\n\u001b[32m    155\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m _cross_merge(\n\u001b[32m    156\u001b[39m         left_df,\n\u001b[32m    157\u001b[39m         right_df,\n\u001b[32m   (...)\u001b[39m\u001b[32m    167\u001b[39m         copy=copy,\n\u001b[32m    168\u001b[39m     )\n\u001b[32m    169\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m170\u001b[39m     op = \u001b[43m_MergeOperation\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    171\u001b[39m \u001b[43m        \u001b[49m\u001b[43mleft_df\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    172\u001b[39m \u001b[43m        \u001b[49m\u001b[43mright_df\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    173\u001b[39m \u001b[43m        \u001b[49m\u001b[43mhow\u001b[49m\u001b[43m=\u001b[49m\u001b[43mhow\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    174\u001b[39m \u001b[43m        \u001b[49m\u001b[43mon\u001b[49m\u001b[43m=\u001b[49m\u001b[43mon\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    175\u001b[39m \u001b[43m        \u001b[49m\u001b[43mleft_on\u001b[49m\u001b[43m=\u001b[49m\u001b[43mleft_on\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    176\u001b[39m \u001b[43m        \u001b[49m\u001b[43mright_on\u001b[49m\u001b[43m=\u001b[49m\u001b[43mright_on\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    177\u001b[39m \u001b[43m        \u001b[49m\u001b[43mleft_index\u001b[49m\u001b[43m=\u001b[49m\u001b[43mleft_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    178\u001b[39m \u001b[43m        \u001b[49m\u001b[43mright_index\u001b[49m\u001b[43m=\u001b[49m\u001b[43mright_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    179\u001b[39m \u001b[43m        \u001b[49m\u001b[43msort\u001b[49m\u001b[43m=\u001b[49m\u001b[43msort\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    180\u001b[39m \u001b[43m        \u001b[49m\u001b[43msuffixes\u001b[49m\u001b[43m=\u001b[49m\u001b[43msuffixes\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    181\u001b[39m \u001b[43m        \u001b[49m\u001b[43mindicator\u001b[49m\u001b[43m=\u001b[49m\u001b[43mindicator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    182\u001b[39m \u001b[43m        \u001b[49m\u001b[43mvalidate\u001b[49m\u001b[43m=\u001b[49m\u001b[43mvalidate\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    183\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    184\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m op.get_result(copy=copy)\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\pandas\\core\\reshape\\merge.py:807\u001b[39m, in \u001b[36m_MergeOperation.__init__\u001b[39m\u001b[34m(self, left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, indicator, validate)\u001b[39m\n\u001b[32m    803\u001b[39m \u001b[38;5;28mself\u001b[39m._validate_tolerance(\u001b[38;5;28mself\u001b[39m.left_join_keys)\n\u001b[32m    805\u001b[39m \u001b[38;5;66;03m# validate the merge keys dtypes. We may need to coerce\u001b[39;00m\n\u001b[32m    806\u001b[39m \u001b[38;5;66;03m# to avoid incompatible dtypes\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m807\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_maybe_coerce_merge_keys\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    809\u001b[39m \u001b[38;5;66;03m# If argument passed to validate,\u001b[39;00m\n\u001b[32m    810\u001b[39m \u001b[38;5;66;03m# check if columns specified as unique\u001b[39;00m\n\u001b[32m    811\u001b[39m \u001b[38;5;66;03m# are in fact unique.\u001b[39;00m\n\u001b[32m    812\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m validate \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\pandas\\core\\reshape\\merge.py:1508\u001b[39m, in \u001b[36m_MergeOperation._maybe_coerce_merge_keys\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m   1502\u001b[39m     \u001b[38;5;66;03m# unless we are merging non-string-like with string-like\u001b[39;00m\n\u001b[32m   1503\u001b[39m     \u001b[38;5;28;01melif\u001b[39;00m (\n\u001b[32m   1504\u001b[39m         inferred_left \u001b[38;5;129;01min\u001b[39;00m string_types \u001b[38;5;129;01mand\u001b[39;00m inferred_right \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m string_types\n\u001b[32m   1505\u001b[39m     ) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[32m   1506\u001b[39m         inferred_right \u001b[38;5;129;01min\u001b[39;00m string_types \u001b[38;5;129;01mand\u001b[39;00m inferred_left \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m string_types\n\u001b[32m   1507\u001b[39m     ):\n\u001b[32m-> \u001b[39m\u001b[32m1508\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(msg)\n\u001b[32m   1510\u001b[39m \u001b[38;5;66;03m# datetimelikes must match exactly\u001b[39;00m\n\u001b[32m   1511\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m needs_i8_conversion(lk.dtype) \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m needs_i8_conversion(rk.dtype):\n",
      "\u001b[31mValueError\u001b[39m: You are trying to merge on object and float64 columns for key 'CellType'. If you wish to proceed you should use pd.concat"
     ]
    }
   ],
   "source": [
    "correlate_atac_with_qc(ATAC, QC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "577b140b",
   "metadata": {},
   "outputs": [
    {
     "ename": "MemoryError",
     "evalue": "Unable to allocate 364. MiB for an array with shape (93, 512595) and data type float64",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mMemoryError\u001b[39m                               Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[2]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m atac_data=\u001b[43mpd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mD:\u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[33;43mUni\u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[33;43m4. Semester\u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[33;43mBioinfo\u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[33;43mdatasets\u001b[39;49m\u001b[38;5;130;43;01m\\\\\u001b[39;49;00m\u001b[33;43mprocessed atac seq data and called peaks.csv\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m      2\u001b[39m \u001b[38;5;28mprint\u001b[39m(mean_atac_data.loc[\u001b[33m'\u001b[39m\u001b[33mLTHSC.34-.BM\u001b[39m\u001b[33m'\u001b[39m])\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\pandas\\io\\parsers\\readers.py:1026\u001b[39m, in \u001b[36mread_csv\u001b[39m\u001b[34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[39m\n\u001b[32m   1013\u001b[39m kwds_defaults = _refine_defaults_read(\n\u001b[32m   1014\u001b[39m     dialect,\n\u001b[32m   1015\u001b[39m     delimiter,\n\u001b[32m   (...)\u001b[39m\u001b[32m   1022\u001b[39m     dtype_backend=dtype_backend,\n\u001b[32m   1023\u001b[39m )\n\u001b[32m   1024\u001b[39m kwds.update(kwds_defaults)\n\u001b[32m-> \u001b[39m\u001b[32m1026\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\pandas\\io\\parsers\\readers.py:626\u001b[39m, in \u001b[36m_read\u001b[39m\u001b[34m(filepath_or_buffer, kwds)\u001b[39m\n\u001b[32m    623\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n\u001b[32m    625\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m parser:\n\u001b[32m--> \u001b[39m\u001b[32m626\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mparser\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\pandas\\io\\parsers\\readers.py:1968\u001b[39m, in \u001b[36mTextFileReader.read\u001b[39m\u001b[34m(self, nrows)\u001b[39m\n\u001b[32m   1965\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m   1966\u001b[39m         new_col_dict = col_dict\n\u001b[32m-> \u001b[39m\u001b[32m1968\u001b[39m     df = \u001b[43mDataFrame\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   1969\u001b[39m \u001b[43m        \u001b[49m\u001b[43mnew_col_dict\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1970\u001b[39m \u001b[43m        \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1971\u001b[39m \u001b[43m        \u001b[49m\u001b[43mindex\u001b[49m\u001b[43m=\u001b[49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1972\u001b[39m \u001b[43m        \u001b[49m\u001b[43mcopy\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43musing_copy_on_write\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1973\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1975\u001b[39m     \u001b[38;5;28mself\u001b[39m._currow += new_rows\n\u001b[32m   1976\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m df\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\pandas\\core\\frame.py:778\u001b[39m, in \u001b[36mDataFrame.__init__\u001b[39m\u001b[34m(self, data, index, columns, dtype, copy)\u001b[39m\n\u001b[32m    772\u001b[39m     mgr = \u001b[38;5;28mself\u001b[39m._init_mgr(\n\u001b[32m    773\u001b[39m         data, axes={\u001b[33m\"\u001b[39m\u001b[33mindex\u001b[39m\u001b[33m\"\u001b[39m: index, \u001b[33m\"\u001b[39m\u001b[33mcolumns\u001b[39m\u001b[33m\"\u001b[39m: columns}, dtype=dtype, copy=copy\n\u001b[32m    774\u001b[39m     )\n\u001b[32m    776\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(data, \u001b[38;5;28mdict\u001b[39m):\n\u001b[32m    777\u001b[39m     \u001b[38;5;66;03m# GH#38939 de facto copy defaults to False only in non-dict cases\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m778\u001b[39m     mgr = \u001b[43mdict_to_mgr\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcopy\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcopy\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtyp\u001b[49m\u001b[43m=\u001b[49m\u001b[43mmanager\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    779\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(data, ma.MaskedArray):\n\u001b[32m    780\u001b[39m     \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mnumpy\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mma\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m mrecords\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\pandas\\core\\internals\\construction.py:503\u001b[39m, in \u001b[36mdict_to_mgr\u001b[39m\u001b[34m(data, index, columns, dtype, typ, copy)\u001b[39m\n\u001b[32m    499\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    500\u001b[39m         \u001b[38;5;66;03m# dtype check to exclude e.g. range objects, scalars\u001b[39;00m\n\u001b[32m    501\u001b[39m         arrays = [x.copy() \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(x, \u001b[33m\"\u001b[39m\u001b[33mdtype\u001b[39m\u001b[33m\"\u001b[39m) \u001b[38;5;28;01melse\u001b[39;00m x \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m arrays]\n\u001b[32m--> \u001b[39m\u001b[32m503\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43marrays_to_mgr\u001b[49m\u001b[43m(\u001b[49m\u001b[43marrays\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtyp\u001b[49m\u001b[43m=\u001b[49m\u001b[43mtyp\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconsolidate\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcopy\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\pandas\\core\\internals\\construction.py:152\u001b[39m, in \u001b[36marrays_to_mgr\u001b[39m\u001b[34m(arrays, columns, index, dtype, verify_integrity, typ, consolidate)\u001b[39m\n\u001b[32m    149\u001b[39m axes = [columns, index]\n\u001b[32m    151\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m typ == \u001b[33m\"\u001b[39m\u001b[33mblock\u001b[39m\u001b[33m\"\u001b[39m:\n\u001b[32m--> \u001b[39m\u001b[32m152\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mcreate_block_manager_from_column_arrays\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    153\u001b[39m \u001b[43m        \u001b[49m\u001b[43marrays\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mconsolidate\u001b[49m\u001b[43m=\u001b[49m\u001b[43mconsolidate\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrefs\u001b[49m\u001b[43m=\u001b[49m\u001b[43mrefs\u001b[49m\n\u001b[32m    154\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    155\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m typ == \u001b[33m\"\u001b[39m\u001b[33marray\u001b[39m\u001b[33m\"\u001b[39m:\n\u001b[32m    156\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m ArrayManager(arrays, [index, columns])\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\pandas\\core\\internals\\managers.py:2144\u001b[39m, in \u001b[36mcreate_block_manager_from_column_arrays\u001b[39m\u001b[34m(arrays, axes, consolidate, refs)\u001b[39m\n\u001b[32m   2142\u001b[39m     raise_construction_error(\u001b[38;5;28mlen\u001b[39m(arrays), arrays[\u001b[32m0\u001b[39m].shape, axes, e)\n\u001b[32m   2143\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m consolidate:\n\u001b[32m-> \u001b[39m\u001b[32m2144\u001b[39m     \u001b[43mmgr\u001b[49m\u001b[43m.\u001b[49m\u001b[43m_consolidate_inplace\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   2145\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m mgr\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\pandas\\core\\internals\\managers.py:1788\u001b[39m, in \u001b[36mBlockManager._consolidate_inplace\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m   1782\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34m_consolidate_inplace\u001b[39m(\u001b[38;5;28mself\u001b[39m) -> \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m   1783\u001b[39m     \u001b[38;5;66;03m# In general, _consolidate_inplace should only be called via\u001b[39;00m\n\u001b[32m   1784\u001b[39m     \u001b[38;5;66;03m#  DataFrame._consolidate_inplace, otherwise we will fail to invalidate\u001b[39;00m\n\u001b[32m   1785\u001b[39m     \u001b[38;5;66;03m#  the DataFrame's _item_cache. The exception is for newly-created\u001b[39;00m\n\u001b[32m   1786\u001b[39m     \u001b[38;5;66;03m#  BlockManager objects not yet attached to a DataFrame.\u001b[39;00m\n\u001b[32m   1787\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m.is_consolidated():\n\u001b[32m-> \u001b[39m\u001b[32m1788\u001b[39m         \u001b[38;5;28mself\u001b[39m.blocks = \u001b[43m_consolidate\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mblocks\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1789\u001b[39m         \u001b[38;5;28mself\u001b[39m._is_consolidated = \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[32m   1790\u001b[39m         \u001b[38;5;28mself\u001b[39m._known_consolidated = \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\pandas\\core\\internals\\managers.py:2269\u001b[39m, in \u001b[36m_consolidate\u001b[39m\u001b[34m(blocks)\u001b[39m\n\u001b[32m   2267\u001b[39m new_blocks: \u001b[38;5;28mlist\u001b[39m[Block] = []\n\u001b[32m   2268\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m (_can_consolidate, dtype), group_blocks \u001b[38;5;129;01min\u001b[39;00m grouper:\n\u001b[32m-> \u001b[39m\u001b[32m2269\u001b[39m     merged_blocks, _ = \u001b[43m_merge_blocks\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   2270\u001b[39m \u001b[43m        \u001b[49m\u001b[38;5;28;43mlist\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mgroup_blocks\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcan_consolidate\u001b[49m\u001b[43m=\u001b[49m\u001b[43m_can_consolidate\u001b[49m\n\u001b[32m   2271\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   2272\u001b[39m     new_blocks = extend_blocks(merged_blocks, new_blocks)\n\u001b[32m   2273\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mtuple\u001b[39m(new_blocks)\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\pandas\\core\\internals\\managers.py:2294\u001b[39m, in \u001b[36m_merge_blocks\u001b[39m\u001b[34m(blocks, dtype, can_consolidate)\u001b[39m\n\u001b[32m   2287\u001b[39m new_values: ArrayLike\n\u001b[32m   2289\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(blocks[\u001b[32m0\u001b[39m].dtype, np.dtype):\n\u001b[32m   2290\u001b[39m     \u001b[38;5;66;03m# error: List comprehension has incompatible type List[Union[ndarray,\u001b[39;00m\n\u001b[32m   2291\u001b[39m     \u001b[38;5;66;03m# ExtensionArray]]; expected List[Union[complex, generic,\u001b[39;00m\n\u001b[32m   2292\u001b[39m     \u001b[38;5;66;03m# Sequence[Union[int, float, complex, str, bytes, generic]],\u001b[39;00m\n\u001b[32m   2293\u001b[39m     \u001b[38;5;66;03m# Sequence[Sequence[Any]], SupportsArray]]\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m2294\u001b[39m     new_values = \u001b[43mnp\u001b[49m\u001b[43m.\u001b[49m\u001b[43mvstack\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43mb\u001b[49m\u001b[43m.\u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mb\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mblocks\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m  \u001b[38;5;66;03m# type: ignore[misc]\u001b[39;00m\n\u001b[32m   2295\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m   2296\u001b[39m     bvals = [blk.values \u001b[38;5;28;01mfor\u001b[39;00m blk \u001b[38;5;129;01min\u001b[39;00m blocks]\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\numpy\\_core\\shape_base.py:292\u001b[39m, in \u001b[36mvstack\u001b[39m\u001b[34m(tup, dtype, casting)\u001b[39m\n\u001b[32m    290\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(arrs, \u001b[38;5;28mtuple\u001b[39m):\n\u001b[32m    291\u001b[39m     arrs = (arrs,)\n\u001b[32m--> \u001b[39m\u001b[32m292\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_nx\u001b[49m\u001b[43m.\u001b[49m\u001b[43mconcatenate\u001b[49m\u001b[43m(\u001b[49m\u001b[43marrs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[32;43m0\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcasting\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcasting\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[31mMemoryError\u001b[39m: Unable to allocate 364. MiB for an array with shape (93, 512595) and data type float64"
     ]
    }
   ],
   "source": [
    "atac_data=pd.read_csv(\"D:\\\\Uni\\\\4. Semester\\\\Bioinfo\\\\datasets\\\\processed atac seq data and called peaks.csv\")\n",
    "print(mean_atac_data.loc['LTHSC.34-.BM'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "613a3fd0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\User\\AppData\\Local\\Temp\\ipykernel_16112\\764912592.py:17: FutureWarning: DataFrame.groupby with axis=1 is deprecated. Do `frame.T.groupby(...)` without axis instead.\n",
      "  mean=peak_signal_log.groupby(peak_signal_log.columns, axis=1).mean()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Zelltyp\n",
      "0.10    1.545527\n",
      "0.11    1.549998\n",
      "0.12    1.521084\n",
      "0.13    1.570888\n",
      "0.14    1.544796\n",
      "dtype: float64\n"
     ]
    }
   ],
   "source": [
    "#Datei einlesen\n",
    "atac_data=pd.read_csv(\"D:\\\\Uni\\\\4. Semester\\\\Bioinfo\\\\datasets\\\\processed atac seq data and called peaks.csv\")\n",
    "\n",
    "#zelltypnamen aus erster zeile nehmen\n",
    "celltype_names=atac_data.iloc[0, 8:]\n",
    "\n",
    "# nur peakdaten benutzen und restliche spalten auslassen\n",
    "score_abgd=atac_data.iloc[1:, 8:].copy().astype(float)\n",
    "\n",
    "# log2 Transformation\n",
    "peak_signal_log = np.log2(score_abgd+1)\n",
    "\n",
    "#Zelltypnamen als Spaltennamen benutzen\n",
    "peak_signal_log.columns=celltype_names.values\n",
    "\n",
    "#mean berechnen pro zelltyp\n",
    "mean=peak_signal_log.groupby(peak_signal_log.columns, axis=1).mean()\n",
    "cell_means=mean.mean(axis=0)\n",
    "\n",
    "#Transponieren\n",
    "mean_T=cell_means.T\n",
    "\n",
    "#dataframe\n",
    "mean_df=mean_T.to_frame(name='mean_accessibility')\n",
    "mean_df.index.name='Zelltyp'\n",
    "\n",
    "\n",
    "print(mean_T.head())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5834d88",
   "metadata": {},
   "source": [
    "### pearson correlation definition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "746ccde0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(512595, 3)\n",
      "(87, 17535)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import pearsonr, linregress\n",
    "from itertools import combinations\n",
    "\n",
    "cell_type=[\"T.4.Th\", \"T.8.Th\", \"T.4.Nve.Sp\"]\n",
    "atac_data=pd.read_csv(\"D:\\\\Uni\\\\4. Semester\\\\Bioinfo\\\\datasets\\\\processed atac seq data and called peaks.csv\", usecols=cell_type)\n",
    "atac_data.index=atac_data.index.astype(str).str.lower().str.replace(\" \",\"\")\n",
    "rna_data=pd.read_csv(\"D:\\\\Uni\\\\4. Semester\\\\Bioinfo\\\\datasets\\\\processed rna seq data.csv\").T\n",
    "qc_metrics=pd.read_excel(\"D:\\\\Uni\\\\4. Semester\\\\Bioinfo\\\\datasets\\\\summary of immune cell populationsprofiled by atac.xlsx\")\n",
    "qc_metrics_cutoff=qc_metrics.iloc[:,1:]\n",
    "qc_metrics_cutoff.index=qc_metrics_cutoff.index.astype(str).str.lower().str.replace(\" \",\"\")\n",
    "\n",
    "print(atac_data.shape)\n",
    "print(rna_data.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d5a1a4e8",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Index' object has no attribute 'index'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mAttributeError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[6]\u001b[39m\u001b[32m, line 79\u001b[39m\n\u001b[32m     72\u001b[39m                 plt.show()\n\u001b[32m     73\u001b[39m                 \u001b[38;5;66;03m# sns.heatmap([[r]], annot=True, cmap='coolwarm', cbar=False, xticklabels=[f\"{df2_name}-{cell2}\"], yticklabels=[f\"{df1_name}-{cell1}\"])\u001b[39;00m\n\u001b[32m     74\u001b[39m                 \u001b[38;5;66;03m# plt.title(f'heatmap-{cell1}-{cell2}')\u001b[39;00m\n\u001b[32m     75\u001b[39m                 \u001b[38;5;66;03m# plt.tight_layout()\u001b[39;00m\n\u001b[32m     76\u001b[39m                 \u001b[38;5;66;03m# plt.show()\u001b[39;00m\n\u001b[32m     77\u001b[39m \u001b[38;5;66;03m# test anwendung\u001b[39;00m\n\u001b[32m     78\u001b[39m \u001b[38;5;66;03m# , df2=rna_data, df2_name='RNA'\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m79\u001b[39m pearson_correlation(plot_type=\u001b[33m'\u001b[39m\u001b[33mheatmap\u001b[39m\u001b[33m'\u001b[39m, cell_list=\u001b[43matac_data\u001b[49m\u001b[43m.\u001b[49m\u001b[43mindex\u001b[49m.tolist(), df1=atac_data, df1_name=\u001b[33m'\u001b[39m\u001b[33mATAC\u001b[39m\u001b[33m'\u001b[39m)\n",
      "\u001b[31mAttributeError\u001b[39m: 'Index' object has no attribute 'index'"
     ]
    }
   ],
   "source": [
    "\n",
    "def pearson_correlation (plot_type, cell_list, df1, df1_name, df2=None, df2_name=None):\n",
    "\n",
    "    '''\n",
    "    df1: first Dataframe\n",
    "    df2: second Dataframe\n",
    "    df1_name: atac/rna/qc/...\n",
    "    df2_name: atac/rna/qc/...\n",
    "    cell1: name of row/celltype of cell 1\n",
    "    cell2: name of row/celltype of cell 2\n",
    "    plot_type: scatterplot/heatmap\n",
    "    '''\n",
    "\n",
    "    # wenn nur ein dataframe gegeben ist, sind beide df1\n",
    "    if df2 is None:\n",
    "        df2=df1\n",
    "        df2_name=df1_name\n",
    "\n",
    "    # falls nur ein df, alle Kombinationen der verschiedenen Zellen miteinander\n",
    "\n",
    "    if df1 is df2:\n",
    "        corr_matrix=pd.DataFrame(index=cell_list, columns=cell_list, dtype=float)\n",
    "        pval_matrix=pd.DataFrame(index=cell_list, columns=cell_list, dtype=float)\n",
    "\n",
    "        for cell1 in cell_list: \n",
    "            for cell2 in cell_list:\n",
    "                row1=pd.to_numeric(df1.loc[cell1], errors='coerce')\n",
    "                row2=pd.to_numeric(df2.loc[cell2], errors='coerce')\n",
    "\n",
    "    # pearson correlation\n",
    "                r, p=pearsonr(row1, row2)\n",
    "                corr_matrix.loc[cell1, cell2]=r\n",
    "                pval_matrix.loc[cell1, cell2]=p\n",
    "\n",
    "    #scatterplot\n",
    "        if plot_type.lower()=='scatterplot':\n",
    "            print(\"Für mehrere Zellen aus einem Datensatz ist es sinnvoller heatmap zu benutzen.\")\n",
    "            \n",
    "    #heatmap\n",
    "        elif plot_type.lower()=='heatmap':\n",
    "            plt.figure(figsize=(3,3))\n",
    "            sns.heatmap(corr_matrix.astype(float), annot=True, cmap='coolwarm')\n",
    "            plt.title(f'heatmap{df1_name}')\n",
    "            plt.tight_layout()\n",
    "            plt.show()\n",
    "    else:\n",
    "        for cell in cell_list: \n",
    "            row1=pd.to_numeric(df1.loc[cell1], errors='coerce')\n",
    "            row2=pd.to_numeric(df2.loc[cell2], errors='coerce')\n",
    "            r, p_wert=pearsonr(row1, row2)\n",
    "            print(f\"pearson-correlation: r= {r:.4f}, p-Wert: {p_wert:.4e}\") \n",
    "\n",
    "            #Scatterplot\n",
    "            if plot_type.lower()=='scatterplot':\n",
    "                plt.figure(figsize=(6,4))\n",
    "                sns.scatterplot(x=row1, y=row2)\n",
    "                slope, intercept, _, _, _ = linregress(row1, row2)\n",
    "                x_vals = pd.Series(sorted(row1))\n",
    "                y_vals = slope * x_vals + intercept\n",
    "                plt.plot(x_vals, y_vals, color='purple', label='Regression')\n",
    "                plt.xlabel(f\"{df1_name}-{cell1}\")\n",
    "                plt.ylabel(f\"{df2_name}-{cell2}\")\n",
    "                plt.title(f'scatterplot-{cell1}-{cell2}')\n",
    "                plt.tight_layout()\n",
    "                plt.show()\n",
    "            \n",
    "            #heatmap\n",
    "            elif plot_type.lower()=='heatmap':\n",
    "                plt.figure(figsize=(3,3))\n",
    "                sns.heatmap(corr_matrix.astype(float), annot=True, cmap='coolwarm')\n",
    "                plt.title(f'heatmap{df1_name}')\n",
    "                plt.tight_layout()\n",
    "                plt.show()\n",
    "                # sns.heatmap([[r]], annot=True, cmap='coolwarm', cbar=False, xticklabels=[f\"{df2_name}-{cell2}\"], yticklabels=[f\"{df1_name}-{cell1}\"])\n",
    "                # plt.title(f'heatmap-{cell1}-{cell2}')\n",
    "                # plt.tight_layout()\n",
    "                # plt.show()\n",
    "# test anwendung\n",
    "# , df2=rna_data, df2_name='RNA'\n",
    "pearson_correlation(plot_type='heatmap', cell_list=atac_data.index.tolist(), df1=atac_data, df1_name='ATAC')\n",
    "\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b0caa431",
   "metadata": {},
   "outputs": [],
   "source": [
    "# heatmap atac mit rna\n",
    "# heatmap rna mit sich selbst\n",
    "#scatterplot von atac mit qc metrics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78174e0a",
   "metadata": {},
   "source": [
    "mean der zelltypen korrelieren mit qc metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "37bc6c1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pearson_correlation_QC_metrics (plot_type, atac_data, qc_data, stat_type='mean', qc_metric_col=None):\n",
    "\n",
    "    '''\n",
    "    df1: first Dataframe (ATAC)\n",
    "    df2: second Dataframe (QC)\n",
    "    qc_metrics_col: welche qc metric soll korreliert werden, spaltenname\n",
    "    plot_type: scatterplot/heatmap\n",
    "    '''\n",
    "    atac_data.columns=atac_data.columns.str.lower().str.replace(\" \", \"\")\n",
    "    qc_data.index=qc_data.index.str.lower().str.replace(\" \", \"\")\n",
    "\n",
    "    if stat_type=='mean':\n",
    "        atac_stat=atac_data.mean(axis=0).rename('ATAC_stat')\n",
    "    elif stat_type=='median':\n",
    "        atac_stat=atac_data.median(axis=0).rename('ATAC_stat')\n",
    "    elif stat_type=='std':\n",
    "        atac_stat=atac_data.std(axis=0).rename('ATAC_stat')\n",
    "\n",
    "    if qc_metric_col not in qc_data.columns:\n",
    "        raise ValueError(f\"QC-Metrik '{qc_metric_col}' nicht gefunden.\")\n",
    "    \n",
    "    qc_metric = qc_data[qc_metric_col]\n",
    "\n",
    "    df=pd.concat([atac_stat, qc_metric], axis=1, join='inner')\n",
    "    df.columns=['ATAC_stat', qc_metric_col]\n",
    "\n",
    "    if len(df)<2:\n",
    "       raise ValueError(\"Nicht genügend überlappende Zelltypen für Korrelation\")\n",
    "\n",
    "\n",
    "    # qc_metric=qc_data.set_index('CellType') [[qc_metric_col]]\n",
    "    # merged=pd.merge(atac_means, qc_metric, left_index=True, right_index=True)\n",
    "    r, p=pearsonr(df['ATAC_stat'], df[qc_metric_col])\n",
    "    print(f\"pearson-correlation ({stat_type} vs {qc_metric_col}): r= {r:.4f}, p-Wert: {p:.4e}\")\n",
    "     \n",
    "            #Scatterplot\n",
    "    if plot_type.lower()=='scatterplot':\n",
    "        plt.figure(figsize=(6,4))\n",
    "        sns.scatterplot(data=merged, x='ATAC_mean', y=qc_metric_col)\n",
    "        slope, intercept, *_ = linregress(merged['ATAC_mean'], merged[qc_metric_col])\n",
    "        x_vals = pd.Series(sorted(merged['ATAC_mean']))\n",
    "        y_vals = slope * x_vals + intercept\n",
    "        plt.plot(x_vals, y_vals, color='purple', label='Regression')\n",
    "        plt.xlabel('ATAC seq mean')\n",
    "        plt.ylabel(qc_metric_col)\n",
    "        plt.title(f'ATAC seq-{qc_metric_col}')\n",
    "        plt.legend()\n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "            \n",
    "            #heatmap\n",
    "    elif plot_type.lower()=='heatmap':\n",
    "        plt.figure(figsize=(3,3))\n",
    "        sns.heatmap([[r]], annot=True, cmap='coolwarm', cbar=True, xticklabels=[qc_metric_col], yticklabels=[\"ATAC_mean\"])\n",
    "        plt.title('Pearson correlation')\n",
    "        plt.tight_layout()\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d3c9e9c6",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "Can only use .str accessor with string values!",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mAttributeError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[13]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43mpearson_correlation_QC_metrics\u001b[49m\u001b[43m(\u001b[49m\u001b[43mplot_type\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mscatterplot\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43matac_data\u001b[49m\u001b[43m=\u001b[49m\u001b[43matac_data\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mqc_data\u001b[49m\u001b[43m=\u001b[49m\u001b[43mqc_metrics\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstat_type\u001b[49m\u001b[43m=\u001b[49m\u001b[43mmean\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mqc_metric_col\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mInputCellNumber\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[12]\u001b[39m\u001b[32m, line 10\u001b[39m, in \u001b[36mpearson_correlation_QC_metrics\u001b[39m\u001b[34m(plot_type, atac_data, qc_data, stat_type, qc_metric_col)\u001b[39m\n\u001b[32m      3\u001b[39m \u001b[38;5;250m\u001b[39m\u001b[33;03m'''\u001b[39;00m\n\u001b[32m      4\u001b[39m \u001b[33;03mdf1: first Dataframe (ATAC)\u001b[39;00m\n\u001b[32m      5\u001b[39m \u001b[33;03mdf2: second Dataframe (QC)\u001b[39;00m\n\u001b[32m      6\u001b[39m \u001b[33;03mqc_metrics_col: welche qc metric soll korreliert werden, spaltenname\u001b[39;00m\n\u001b[32m      7\u001b[39m \u001b[33;03mplot_type: scatterplot/heatmap\u001b[39;00m\n\u001b[32m      8\u001b[39m \u001b[33;03m'''\u001b[39;00m\n\u001b[32m      9\u001b[39m atac_data.columns=atac_data.columns.str.lower().str.replace(\u001b[33m\"\u001b[39m\u001b[33m \u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m---> \u001b[39m\u001b[32m10\u001b[39m qc_data.index=\u001b[43mqc_data\u001b[49m\u001b[43m.\u001b[49m\u001b[43mindex\u001b[49m\u001b[43m.\u001b[49m\u001b[43mstr\u001b[49m.lower().str.replace(\u001b[33m\"\u001b[39m\u001b[33m \u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m     12\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m stat_type==\u001b[33m'\u001b[39m\u001b[33mmean\u001b[39m\u001b[33m'\u001b[39m:\n\u001b[32m     13\u001b[39m     atac_stat=atac_data.mean(axis=\u001b[32m0\u001b[39m).rename(\u001b[33m'\u001b[39m\u001b[33mATAC_stat\u001b[39m\u001b[33m'\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\pandas\\core\\accessor.py:224\u001b[39m, in \u001b[36mCachedAccessor.__get__\u001b[39m\u001b[34m(self, obj, cls)\u001b[39m\n\u001b[32m    221\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m obj \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m    222\u001b[39m     \u001b[38;5;66;03m# we're accessing the attribute of the class, i.e., Dataset.geo\u001b[39;00m\n\u001b[32m    223\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m._accessor\n\u001b[32m--> \u001b[39m\u001b[32m224\u001b[39m accessor_obj = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_accessor\u001b[49m\u001b[43m(\u001b[49m\u001b[43mobj\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    225\u001b[39m \u001b[38;5;66;03m# Replace the property with the accessor object. Inspired by:\u001b[39;00m\n\u001b[32m    226\u001b[39m \u001b[38;5;66;03m# https://www.pydanny.com/cached-property.html\u001b[39;00m\n\u001b[32m    227\u001b[39m \u001b[38;5;66;03m# We need to use object.__setattr__ because we overwrite __setattr__ on\u001b[39;00m\n\u001b[32m    228\u001b[39m \u001b[38;5;66;03m# NDFrame\u001b[39;00m\n\u001b[32m    229\u001b[39m \u001b[38;5;28mobject\u001b[39m.\u001b[34m__setattr__\u001b[39m(obj, \u001b[38;5;28mself\u001b[39m._name, accessor_obj)\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\pandas\\core\\strings\\accessor.py:191\u001b[39m, in \u001b[36mStringMethods.__init__\u001b[39m\u001b[34m(self, data)\u001b[39m\n\u001b[32m    188\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34m__init__\u001b[39m(\u001b[38;5;28mself\u001b[39m, data) -> \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m    189\u001b[39m     \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mpandas\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mcore\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01marrays\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mstring_\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m StringDtype\n\u001b[32m--> \u001b[39m\u001b[32m191\u001b[39m     \u001b[38;5;28mself\u001b[39m._inferred_dtype = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_validate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    192\u001b[39m     \u001b[38;5;28mself\u001b[39m._is_categorical = \u001b[38;5;28misinstance\u001b[39m(data.dtype, CategoricalDtype)\n\u001b[32m    193\u001b[39m     \u001b[38;5;28mself\u001b[39m._is_string = \u001b[38;5;28misinstance\u001b[39m(data.dtype, StringDtype)\n",
      "\u001b[36mFile \u001b[39m\u001b[32md:\\Uni\\4. Semester\\Bioinfo\\abT-gdT-cells-Team3\\abT-gdT-cells-Team3\\.venv\\Lib\\site-packages\\pandas\\core\\strings\\accessor.py:245\u001b[39m, in \u001b[36mStringMethods._validate\u001b[39m\u001b[34m(data)\u001b[39m\n\u001b[32m    242\u001b[39m inferred_dtype = lib.infer_dtype(values, skipna=\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[32m    244\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m inferred_dtype \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m allowed_types:\n\u001b[32m--> \u001b[39m\u001b[32m245\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mAttributeError\u001b[39;00m(\u001b[33m\"\u001b[39m\u001b[33mCan only use .str accessor with string values!\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m    246\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m inferred_dtype\n",
      "\u001b[31mAttributeError\u001b[39m: Can only use .str accessor with string values!"
     ]
    }
   ],
   "source": [
    "pearson_correlation_QC_metrics(plot_type='scatterplot', atac_data=atac_data, qc_data=qc_metrics, stat_type=mean, qc_metric_col='InputCellNumber')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "446d10d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180]\n"
     ]
    }
   ],
   "source": [
    "qc_metrics=pd.read_excel(\"D:\\\\Uni\\\\4. Semester\\\\Bioinfo\\\\datasets\\\\summary of immune cell populationsprofiled by atac.xlsx\")\n",
    "qc_metrics_cutoff=qc_metrics.iloc[:,1:]\n",
    "\n",
    "print(list(qc_metrics_cutoff.index))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
