{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f0553079",
   "metadata": {},
   "source": [
    "# Creation of UMAP to identify batch effects"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "233fffa1",
   "metadata": {},
   "source": [
    "Import Packages and Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a70cbbfb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/envs/dataanalysisenv/lib/python3.13/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "#import packages\n",
    "import pandas as pd\n",
    "import umap \n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Pia Kentschke\n",
    "ATAC_seq_path = \"/Users/piakentschke/Documents/Uni/Data Analysis/ImmGenATAC18_AllOCRsInfo.csv\"\n",
    "ATAC_scores_path = \"/Users/piakentschke/Documents/Uni/Data Analysis/ATAC_scores.csv\"\n",
    "RNA_seq_path = \"/Users/piakentschke/Documents/Uni/Data Analysis/RNA-seq_data.csv\"\n",
    "Transcription_exons_path = \"/Users/piakentschke/Documents/Uni/Data Analysis/refFlat.txt\"\n",
    "Cell_population_qc_path = \"/Users/piakentschke/Documents/Uni/Data Analysis/mmc1.xlsx\"\n",
    "Voluntary_path = \"/Users/piakentschke/Documents/Uni/Data Analysis/ImmGenATAC18_AllTFmotifsInOCRs.txt\"\n",
    "\n",
    "\n",
    "ATAC_scores = pd.read_csv(ATAC_scores_path, sep=\",\", keep_default_na=False)\n",
    "RNA_seq = pd.read_csv(RNA_seq_path)\n",
    "ATAC_seq = pd.read_csv(ATAC_seq_path, sep=\",\", keep_default_na=False)\n",
    "\n",
    "#drop peaks with infinite -log10 p values\n",
    "ATAC_seq = ATAC_seq[ATAC_seq['_-log10_bestPvalue'] != np.inf]\n",
    "\n",
    "#define threshold (p = 0.05)\n",
    "threshold_pval = 1.3\n",
    "\n",
    "# Filter ATAC-seq peaks based on logp value threshold \n",
    "ATAC_cleaned = ATAC_seq[ATAC_seq[\"_-log10_bestPvalue\"] >= threshold_pval]\n",
    "\n",
    "\n",
    "# Create data subsets\n",
    "col_ATAC_seq_ab_act_gd_T = ['LTHSC.34-.BM', 'LTHSC.34+.BM', 'STHSC.150-.BM', 'MPP4.135+.BM','preT.DN1.Th', 'preT.DN2a.Th', 'preT.DN2b.Th',\n",
    "    'preT.DN3.Th', 'T.DN4.Th', 'T.ISP.Th', 'T.DP.Th', 'T.4.Th', 'T.8.Th',\n",
    "    'T.4.Nve.Sp', 'T.4.Nve.Fem.Sp', 'T.8.Nve.Sp', 'T.4.Sp.aCD3+CD40.18hr',\n",
    "    'Treg.4.FP3+.Nrplo.Co', 'Treg.4.25hi.Sp', 'T8.TN.P14.Sp',\n",
    "    'T8.IEL.LCMV.d7.SI', 'T8.TE.LCMV.d7.Sp', 'T8.MP.LCMV.d7.Sp',\n",
    "    'T8.Tcm.LCMV.d180.Sp', 'T8.Tem.LCMV.d180.Sp', 'NKT.Sp',\n",
    "    'NKT.Sp.LPS.3hr', 'NKT.Sp.LPS.18hr', 'NKT.Sp.LPS.3d',\n",
    "    'Tgd.g2+d17.24a+.Th', 'Tgd.g2+d17.LN', 'Tgd.g2+d1.24a+.Th',\n",
    "    'Tgd.g2+d1.LN', 'Tgd.g1.1+d1.24a+.Th', 'Tgd.g1.1+d1.LN', 'Tgd.Sp']\n",
    "\n",
    "col_ATAC_seq_ab_act_T = ['LTHSC.34-.BM', 'LTHSC.34+.BM', 'STHSC.150-.BM', 'MPP4.135+.BM','preT.DN1.Th', 'preT.DN2a.Th', 'preT.DN2b.Th',\n",
    "    'preT.DN3.Th', 'T.DN4.Th', 'T.ISP.Th', 'T.DP.Th', 'T.4.Th', 'T.8.Th',\n",
    "    'T.4.Nve.Sp', 'T.4.Nve.Fem.Sp', 'T.8.Nve.Sp', 'T.4.Sp.aCD3+CD40.18hr',\n",
    "    'Treg.4.FP3+.Nrplo.Co', 'Treg.4.25hi.Sp', 'T8.TN.P14.Sp',\n",
    "    'T8.IEL.LCMV.d7.SI', 'T8.TE.LCMV.d7.Sp', 'T8.MP.LCMV.d7.Sp',\n",
    "    'T8.Tcm.LCMV.d180.Sp', 'T8.Tem.LCMV.d180.Sp', 'NKT.Sp',\n",
    "    'NKT.Sp.LPS.3hr', 'NKT.Sp.LPS.18hr', 'NKT.Sp.LPS.3d']\n",
    "\n",
    "col_ATAC_seq_gd_T = ['LTHSC.34-.BM', 'LTHSC.34+.BM', 'STHSC.150-.BM', 'MPP4.135+.BM','Tgd.g2+d17.24a+.Th', 'Tgd.g2+d17.LN', 'Tgd.g2+d1.24a+.Th',\n",
    "    'Tgd.g2+d1.LN', 'Tgd.g1.1+d1.24a+.Th', 'Tgd.g1.1+d1.LN', 'Tgd.Sp']\n",
    "\n",
    "col_ATAC_seq_ab_T = ['LTHSC.34-.BM', 'LTHSC.34+.BM', 'STHSC.150-.BM', 'MPP4.135+.BM','preT.DN1.Th', 'preT.DN2a.Th', 'preT.DN2b.Th',\n",
    "    'preT.DN3.Th', 'T.DN4.Th', 'T.ISP.Th', 'T.DP.Th', 'T.4.Th', 'T.8.Th',\n",
    "    'T.4.Nve.Sp', 'T.4.Nve.Fem.Sp', 'T.8.Nve.Sp', 'T.4.Sp.aCD3+CD40.18hr',\n",
    "    'Treg.4.FP3+.Nrplo.Co', 'Treg.4.25hi.Sp']\n",
    "\n",
    "ATAC_seq_ab_act_gd_T = ATAC_cleaned[col_ATAC_seq_ab_act_gd_T]\n",
    "ATAC_seq_gd_T = ATAC_seq[col_ATAC_seq_gd_T]\n",
    "ATAC_seq_ab_act_T = ATAC_seq[col_ATAC_seq_ab_act_T]\n",
    "ATAC_seq_ab_T = ATAC_seq[col_ATAC_seq_ab_T]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "71c9de0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# columns sorted by cell lineage\n",
    "Stem_prog_cols = ['LTHSC.34-.BM', 'LTHSC.34+.BM', 'STHSC.150-.BM', 'MPP4.135+.BM', 'MPP3.48+.BM']\n",
    "Tab_cols = ['preT.DN1.Th', 'preT.DN2a.Th', 'preT.DN2b.Th', 'preT.DN3.Th', 'T.DN4.Th', 'T.ISP.Th', 'T.DP.Th', 'T.4.Th', 'T.8.Th','T.4.Nve.Sp', \n",
    "        'T.4.Nve.Fem.Sp', 'T.8.Nve.Sp', 'T.4.Sp.aCD3+CD40.18hr', 'Treg.4.FP3+.Nrplo.Co', 'Treg.4.25hi.Sp','T8.TN.P14.Sp', 'T8.IEL.LCMV.d7.SI',\n",
    "        'T8.TE.LCMV.d7.Sp', 'T8.MP.LCMV.d7.Sp','T8.Tcm.LCMV.d180.Sp', 'T8.Tem.LCMV.d180.Sp', 'NKT.Sp'] \n",
    "\n",
    "Tact_cols = ['NKT.Sp.LPS.3hr', 'NKT.Sp.LPS.18hr', 'NKT.Sp.LPS.3d']\n",
    "\n",
    "Tgd_cols = ['Tgd.g2+d17.24a+.Th', 'Tgd.g2+d17.LN', 'Tgd.g2+d1.24a+.Th', 'Tgd.g2+d1.LN', 'Tgd.g1.1+d1.24a+.Th', 'Tgd.g1.1+d1.LN', 'Tgd.Sp']\n",
    "\n",
    "B_cols = ['proB.CLP.BM', 'proB.FrA.BM', 'proB.FrBC.BM', 'preB.FrD.BM', 'B.FrE.BM',\n",
    "       'B1b.PC', 'B.T1.Sp', 'B.T2.Sp', 'B.T3.Sp', 'B.Sp', 'B.Fem.Sp',\n",
    "       'B.MZ.Sp', 'B.Fo.Sp', 'B.mem.Sp', 'B.GC.CB.Sp', 'B.GC.CC.Sp', 'B.PB.Sp',\n",
    "       'B.PC.Sp', 'B.PC.BM']\n",
    "\n",
    "ILC_cols = ['NK.27+11b-.BM', 'NK.27+11b+.BM', 'NK.27-11b+.BM', 'NK.27+11b-.Sp','NK.27+11b+.Sp', 'NK.27-11b+.Sp', 'ILC2.SI', 'ILC3.NKp46-CCR6-.SI',\n",
    "       'ILC3.NKp46+.SI', 'ILC3.CCR6+.SI']\n",
    "\n",
    "my_cols = ['GN.BM', 'GN.Sp','GN.Thio.PC', 'MF.pIC.Alv.Lu', 'MF.226+II+480lo.PC', 'MF.ICAM+480hi.PC', 'MF.RP.Sp', 'MF.LP.SI', 'MF.Alv.Lu', \n",
    "        'MF.microglia.CNS', 'MF.PC', 'MF.Fem.PC']\n",
    "\n",
    "mono_cols = ['Mo.6C+II-.Bl', 'Mo.6C-II-.Bl']\n",
    "\n",
    "dc_cols = ['DC.4+.Sp', 'DC.8+.Sp', 'DC.pDC.Sp', 'DC.103+11b+.SI', 'DC.103+11b-.SI']\n",
    "\n",
    "stroma_cols = ['FRC.SLN','IAP.SLN', 'BEC.SLN', 'LEC.SLN', 'Ep.MEChi.Th']\n",
    "\n",
    "cell_lineages = {\n",
    "    'Stem_prog': Stem_prog_cols,\n",
    "    'Tab': Tab_cols,\n",
    "    'Tact': Tact_cols,\n",
    "    'Tgd': Tgd_cols,\n",
    "    'B': B_cols,\n",
    "    'ILC': ILC_cols,\n",
    "    'Myeloid': my_cols,\n",
    "    'Mono': mono_cols,\n",
    "    'DC': dc_cols,\n",
    "    'Stroma': stroma_cols\n",
    "}\n",
    "\n",
    "# Flatten the mapping from cell type -> columns into column -> cell type\n",
    "column_to_lineage = {}\n",
    "for lineage, cols in cell_lineages.items():\n",
    "    for col in cols:\n",
    "        column_to_lineage[col] = lineage\n",
    "\n",
    "# Create a list of lineage labels\n",
    "sample_columns = ATAC_scores.columns[:] \n",
    "\n",
    "# Build a label list in the correct order\n",
    "lineage_labels = [column_to_lineage.get(col, 'Unknown') for col in sample_columns]\n",
    "\n",
    "# transpose the ATAC_scores DataFrame to have samples as rows and peaks as columns\n",
    "umap_data = ATAC_scores.iloc[:, :].T  \n",
    "\n",
    "# Apply UMAP\n",
    "umap_embedding = umap.UMAP(n_neighbors=15, min_dist=0.1, metric='euclidean').fit_transform(umap_data.iloc[1:, :])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b6bea89f",
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "'c' argument has 91 elements, which is inconsistent with 'x' and 'y' with size 90.",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mValueError\u001b[39m                                Traceback (most recent call last)",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/dataanalysisenv/lib/python3.13/site-packages/matplotlib/axes/_axes.py:4725\u001b[39m, in \u001b[36mAxes._parse_scatter_color_args\u001b[39m\u001b[34m(c, edgecolors, kwargs, xsize, get_next_color_func)\u001b[39m\n\u001b[32m   4724\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:  \u001b[38;5;66;03m# Is 'c' acceptable as PathCollection facecolors?\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m4725\u001b[39m     colors = \u001b[43mmcolors\u001b[49m\u001b[43m.\u001b[49m\u001b[43mto_rgba_array\u001b[49m\u001b[43m(\u001b[49m\u001b[43mc\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   4726\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mTypeError\u001b[39;00m, \u001b[38;5;167;01mValueError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/dataanalysisenv/lib/python3.13/site-packages/matplotlib/colors.py:514\u001b[39m, in \u001b[36mto_rgba_array\u001b[39m\u001b[34m(c, alpha)\u001b[39m\n\u001b[32m    513\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m514\u001b[39m     rgba = np.array([\u001b[43mto_rgba\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcc\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mfor\u001b[39;00m cc \u001b[38;5;129;01min\u001b[39;00m c])\n\u001b[32m    516\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m alpha \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/dataanalysisenv/lib/python3.13/site-packages/matplotlib/colors.py:316\u001b[39m, in \u001b[36mto_rgba\u001b[39m\u001b[34m(c, alpha)\u001b[39m\n\u001b[32m    315\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m rgba \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:  \u001b[38;5;66;03m# Suppress exception chaining of cache lookup failure.\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m316\u001b[39m     rgba = \u001b[43m_to_rgba_no_colorcycle\u001b[49m\u001b[43m(\u001b[49m\u001b[43mc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43malpha\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    317\u001b[39m     \u001b[38;5;28;01mtry\u001b[39;00m:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/dataanalysisenv/lib/python3.13/site-packages/matplotlib/colors.py:400\u001b[39m, in \u001b[36m_to_rgba_no_colorcycle\u001b[39m\u001b[34m(c, alpha)\u001b[39m\n\u001b[32m    399\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m np.iterable(c):\n\u001b[32m--> \u001b[39m\u001b[32m400\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mInvalid RGBA argument: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00morig_c\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[33m\"\u001b[39m)\n\u001b[32m    401\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(c) \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m [\u001b[32m3\u001b[39m, \u001b[32m4\u001b[39m]:\n",
      "\u001b[31mValueError\u001b[39m: Invalid RGBA argument: np.float64(10.0)",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[31mValueError\u001b[39m                                Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[19]\u001b[39m\u001b[32m, line 12\u001b[39m\n\u001b[32m     10\u001b[39m \u001b[38;5;66;03m# UMAP plot\u001b[39;00m\n\u001b[32m     11\u001b[39m plt.figure(figsize=(\u001b[32m8\u001b[39m, \u001b[32m6\u001b[39m))\n\u001b[32m---> \u001b[39m\u001b[32m12\u001b[39m \u001b[43mplt\u001b[49m\u001b[43m.\u001b[49m\u001b[43mscatter\u001b[49m\u001b[43m(\u001b[49m\u001b[43mumap_embedding\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[32;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mumap_embedding\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[32;43m1\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mc\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcolors\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcmap\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mSpectral\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43ms\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m5\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43malpha\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m0.8\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m     13\u001b[39m plt.title(\u001b[33m'\u001b[39m\u001b[33mUMAP of ATAC Samples by Cell Lineage\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m     14\u001b[39m plt.xlabel(\u001b[33m'\u001b[39m\u001b[33mUMAP1\u001b[39m\u001b[33m'\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/dataanalysisenv/lib/python3.13/site-packages/matplotlib/_api/deprecation.py:453\u001b[39m, in \u001b[36mmake_keyword_only.<locals>.wrapper\u001b[39m\u001b[34m(*args, **kwargs)\u001b[39m\n\u001b[32m    447\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) > name_idx:\n\u001b[32m    448\u001b[39m     warn_deprecated(\n\u001b[32m    449\u001b[39m         since, message=\u001b[33m\"\u001b[39m\u001b[33mPassing the \u001b[39m\u001b[38;5;132;01m%(name)s\u001b[39;00m\u001b[33m \u001b[39m\u001b[38;5;132;01m%(obj_type)s\u001b[39;00m\u001b[33m \u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    450\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mpositionally is deprecated since Matplotlib \u001b[39m\u001b[38;5;132;01m%(since)s\u001b[39;00m\u001b[33m; the \u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    451\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mparameter will become keyword-only in \u001b[39m\u001b[38;5;132;01m%(removal)s\u001b[39;00m\u001b[33m.\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m    452\u001b[39m         name=name, obj_type=\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mparameter of \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfunc.\u001b[34m__name__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m()\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m--> \u001b[39m\u001b[32m453\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/dataanalysisenv/lib/python3.13/site-packages/matplotlib/pyplot.py:3939\u001b[39m, in \u001b[36mscatter\u001b[39m\u001b[34m(x, y, s, c, marker, cmap, norm, vmin, vmax, alpha, linewidths, edgecolors, colorizer, plotnonfinite, data, **kwargs)\u001b[39m\n\u001b[32m   3919\u001b[39m \u001b[38;5;129m@_copy_docstring_and_deprecators\u001b[39m(Axes.scatter)\n\u001b[32m   3920\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mscatter\u001b[39m(\n\u001b[32m   3921\u001b[39m     x: \u001b[38;5;28mfloat\u001b[39m | ArrayLike,\n\u001b[32m   (...)\u001b[39m\u001b[32m   3937\u001b[39m     **kwargs,\n\u001b[32m   3938\u001b[39m ) -> PathCollection:\n\u001b[32m-> \u001b[39m\u001b[32m3939\u001b[39m     __ret = \u001b[43mgca\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m.\u001b[49m\u001b[43mscatter\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   3940\u001b[39m \u001b[43m        \u001b[49m\u001b[43mx\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3941\u001b[39m \u001b[43m        \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3942\u001b[39m \u001b[43m        \u001b[49m\u001b[43ms\u001b[49m\u001b[43m=\u001b[49m\u001b[43ms\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3943\u001b[39m \u001b[43m        \u001b[49m\u001b[43mc\u001b[49m\u001b[43m=\u001b[49m\u001b[43mc\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3944\u001b[39m \u001b[43m        \u001b[49m\u001b[43mmarker\u001b[49m\u001b[43m=\u001b[49m\u001b[43mmarker\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3945\u001b[39m \u001b[43m        \u001b[49m\u001b[43mcmap\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcmap\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3946\u001b[39m \u001b[43m        \u001b[49m\u001b[43mnorm\u001b[49m\u001b[43m=\u001b[49m\u001b[43mnorm\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3947\u001b[39m \u001b[43m        \u001b[49m\u001b[43mvmin\u001b[49m\u001b[43m=\u001b[49m\u001b[43mvmin\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3948\u001b[39m \u001b[43m        \u001b[49m\u001b[43mvmax\u001b[49m\u001b[43m=\u001b[49m\u001b[43mvmax\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3949\u001b[39m \u001b[43m        \u001b[49m\u001b[43malpha\u001b[49m\u001b[43m=\u001b[49m\u001b[43malpha\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3950\u001b[39m \u001b[43m        \u001b[49m\u001b[43mlinewidths\u001b[49m\u001b[43m=\u001b[49m\u001b[43mlinewidths\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3951\u001b[39m \u001b[43m        \u001b[49m\u001b[43medgecolors\u001b[49m\u001b[43m=\u001b[49m\u001b[43medgecolors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3952\u001b[39m \u001b[43m        \u001b[49m\u001b[43mcolorizer\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcolorizer\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3953\u001b[39m \u001b[43m        \u001b[49m\u001b[43mplotnonfinite\u001b[49m\u001b[43m=\u001b[49m\u001b[43mplotnonfinite\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3954\u001b[39m \u001b[43m        \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43m(\u001b[49m\u001b[43m{\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mdata\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m}\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mif\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mis\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;129;43;01mnot\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[38;5;28;43;01melse\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43m{\u001b[49m\u001b[43m}\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3955\u001b[39m \u001b[43m        \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   3956\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   3957\u001b[39m     sci(__ret)\n\u001b[32m   3958\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m __ret\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/dataanalysisenv/lib/python3.13/site-packages/matplotlib/_api/deprecation.py:453\u001b[39m, in \u001b[36mmake_keyword_only.<locals>.wrapper\u001b[39m\u001b[34m(*args, **kwargs)\u001b[39m\n\u001b[32m    447\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) > name_idx:\n\u001b[32m    448\u001b[39m     warn_deprecated(\n\u001b[32m    449\u001b[39m         since, message=\u001b[33m\"\u001b[39m\u001b[33mPassing the \u001b[39m\u001b[38;5;132;01m%(name)s\u001b[39;00m\u001b[33m \u001b[39m\u001b[38;5;132;01m%(obj_type)s\u001b[39;00m\u001b[33m \u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    450\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mpositionally is deprecated since Matplotlib \u001b[39m\u001b[38;5;132;01m%(since)s\u001b[39;00m\u001b[33m; the \u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    451\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mparameter will become keyword-only in \u001b[39m\u001b[38;5;132;01m%(removal)s\u001b[39;00m\u001b[33m.\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m    452\u001b[39m         name=name, obj_type=\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mparameter of \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfunc.\u001b[34m__name__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m()\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m--> \u001b[39m\u001b[32m453\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/dataanalysisenv/lib/python3.13/site-packages/matplotlib/__init__.py:1521\u001b[39m, in \u001b[36m_preprocess_data.<locals>.inner\u001b[39m\u001b[34m(ax, data, *args, **kwargs)\u001b[39m\n\u001b[32m   1518\u001b[39m \u001b[38;5;129m@functools\u001b[39m.wraps(func)\n\u001b[32m   1519\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34minner\u001b[39m(ax, *args, data=\u001b[38;5;28;01mNone\u001b[39;00m, **kwargs):\n\u001b[32m   1520\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m data \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m-> \u001b[39m\u001b[32m1521\u001b[39m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   1522\u001b[39m \u001b[43m            \u001b[49m\u001b[43max\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1523\u001b[39m \u001b[43m            \u001b[49m\u001b[43m*\u001b[49m\u001b[38;5;28;43mmap\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mcbook\u001b[49m\u001b[43m.\u001b[49m\u001b[43msanitize_sequence\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   1524\u001b[39m \u001b[43m            \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43m{\u001b[49m\u001b[43mk\u001b[49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mcbook\u001b[49m\u001b[43m.\u001b[49m\u001b[43msanitize_sequence\u001b[49m\u001b[43m(\u001b[49m\u001b[43mv\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mfor\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mv\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;129;43;01min\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mitems\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m}\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1526\u001b[39m     bound = new_sig.bind(ax, *args, **kwargs)\n\u001b[32m   1527\u001b[39m     auto_label = (bound.arguments.get(label_namer)\n\u001b[32m   1528\u001b[39m                   \u001b[38;5;129;01mor\u001b[39;00m bound.kwargs.get(label_namer))\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/dataanalysisenv/lib/python3.13/site-packages/matplotlib/axes/_axes.py:4918\u001b[39m, in \u001b[36mAxes.scatter\u001b[39m\u001b[34m(self, x, y, s, c, marker, cmap, norm, vmin, vmax, alpha, linewidths, edgecolors, colorizer, plotnonfinite, **kwargs)\u001b[39m\n\u001b[32m   4915\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m edgecolors \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m   4916\u001b[39m     orig_edgecolor = kwargs.get(\u001b[33m'\u001b[39m\u001b[33medgecolor\u001b[39m\u001b[33m'\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[32m   4917\u001b[39m c, colors, edgecolors = \\\n\u001b[32m-> \u001b[39m\u001b[32m4918\u001b[39m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_parse_scatter_color_args\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m   4919\u001b[39m \u001b[43m        \u001b[49m\u001b[43mc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43medgecolors\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mx\u001b[49m\u001b[43m.\u001b[49m\u001b[43msize\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m   4920\u001b[39m \u001b[43m        \u001b[49m\u001b[43mget_next_color_func\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_get_patches_for_fill\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_next_color\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   4922\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m plotnonfinite \u001b[38;5;129;01mand\u001b[39;00m colors \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m   4923\u001b[39m     c = np.ma.masked_invalid(c)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/dataanalysisenv/lib/python3.13/site-packages/matplotlib/axes/_axes.py:4731\u001b[39m, in \u001b[36mAxes._parse_scatter_color_args\u001b[39m\u001b[34m(c, edgecolors, kwargs, xsize, get_next_color_func)\u001b[39m\n\u001b[32m   4729\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m   4730\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m valid_shape:\n\u001b[32m-> \u001b[39m\u001b[32m4731\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m invalid_shape_exception(c.size, xsize) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01merr\u001b[39;00m\n\u001b[32m   4732\u001b[39m     \u001b[38;5;66;03m# Both the mapping *and* the RGBA conversion failed: pretty\u001b[39;00m\n\u001b[32m   4733\u001b[39m     \u001b[38;5;66;03m# severe failure => one may appreciate a verbose feedback.\u001b[39;00m\n\u001b[32m   4734\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[32m   4735\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33m'\u001b[39m\u001b[33mc\u001b[39m\u001b[33m'\u001b[39m\u001b[33m argument must be a color, a sequence of colors, \u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m   4736\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mor a sequence of numbers, not \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mc\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[33m\"\u001b[39m) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01merr\u001b[39;00m\n",
      "\u001b[31mValueError\u001b[39m: 'c' argument has 91 elements, which is inconsistent with 'x' and 'y' with size 90."
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAqoAAAH/CAYAAACfLv+zAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjAsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvlHJYcgAAAAlwSFlzAAAPYQAAD2EBqD+naQAAH4xJREFUeJzt3X9s1fW9+PFXodCq97aLMCsIMtjVjY1cdymBUS9Z5tUaNC7c7EYWb0S9mqzZdhF63R2MGx1kSbPdzNy5CW4TNEvQS/BX/KPX0T/uRRDuD7hlWQaJi3AtzFZSjC3qVgQ+3z/80u+3t8VxDm15SR+P5Pxx3nu/T99n79U99zmnn1UURVEEAAAkM+5CbwAAAIYiVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBIqeRQffnll+O2226LqVOnRkVFRbzwwgt/cM327dujvr4+qqurY9asWfHYY4+Vs1cAAMaQkkP13Xffjeuuuy5+/OMfn9P8Q4cOxS233BKLFi2K9vb2+Pa3vx3Lly+PZ599tuTNAgAwdlQURVGUvbiiIp5//vlYsmTJWed861vfihdffDEOHDjQP9bU1BS//OUvY/fu3eX+aAAALnKVI/0Ddu/eHY2NjQPGbr755ti4cWO8//77MWHChEFr+vr6oq+vr//56dOn46233opJkyZFRUXFSG8ZAIASFUURx48fj6lTp8a4ccPzZ1AjHqpdXV1RV1c3YKyuri5OnjwZ3d3dMWXKlEFrWlpaYu3atSO9NQAAhtnhw4dj2rRpw/JaIx6qETHoKuiZbxuc7ero6tWro7m5uf95T09PXH311XH48OGoqakZuY0CAFCW3t7emD59evzxH//xsL3miIfqlVdeGV1dXQPGjh49GpWVlTFp0qQh11RVVUVVVdWg8ZqaGqEKAJDYcH5Nc8Tvo7pw4cJoa2sbMLZt27aYN2/ekN9PBQCAiDJC9Z133ol9+/bFvn37IuKD20/t27cvOjo6IuKDj+2XLVvWP7+pqSlef/31aG5ujgMHDsSmTZti48aN8cADDwzPOwAA4KJU8kf/e/bsiS9+8Yv9z898l/Suu+6KJ598Mjo7O/ujNSJi5syZ0draGitXroxHH300pk6dGo888kh8+ctfHobtAwBwsTqv+6iOlt7e3qitrY2enh7fUQUASGgkem3Ev6MKAADlEKoAAKQkVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAICWhCgBASkIVAICUhCoAACkJVQAAUhKqAACkJFQBAEhJqAIAkJJQBQAgJaEKAEBKQhUAgJSEKgAAKQlVAABSEqoAAKQkVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAICWhCgBASkIVAICUhCoAACkJVQAAUhKqAACkJFQBAEhJqAIAkJJQBQAgJaEKAEBKQhUAgJSEKgAAKQlVAABSEqoAAKQkVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAICWhCgBASkIVAICUhCoAACkJVQAAUhKqAACkJFQBAEhJqAIAkJJQBQAgJaEKAEBKQhUAgJSEKgAAKQlVAABSEqoAAKQkVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFIqK1TXr18fM2fOjOrq6qivr48dO3Z86PzNmzfHddddF5deemlMmTIl7rnnnjh27FhZGwYAYGwoOVS3bNkSK1asiDVr1kR7e3ssWrQoFi9eHB0dHUPO37lzZyxbtizuvffe+PWvfx1bt26N//qv/4r77rvvvDcPAMDFq+RQffjhh+Pee++N++67L2bPnh3/9E//FNOnT48NGzYMOf/f//3f4xOf+EQsX748Zs6cGX/+538eX/3qV2PPnj3nvXkAAC5eJYXqiRMnYu/evdHY2DhgvLGxMXbt2jXkmoaGhjhy5Ei0trZGURTx5ptvxjPPPBO33npr+bsGAOCiV1Kodnd3x6lTp6Kurm7AeF1dXXR1dQ25pqGhITZv3hxLly6NiRMnxpVXXhkf+9jH4kc/+tFZf05fX1/09vYOeAAAMLaU9cdUFRUVA54XRTFo7Iz9+/fH8uXL48EHH4y9e/fGSy+9FIcOHYqmpqazvn5LS0vU1tb2P6ZPn17ONgEA+AirKIqiONfJJ06ciEsvvTS2bt0af/mXf9k/fv/998e+ffti+/btg9bceeed8fvf/z62bt3aP7Zz585YtGhRvPHGGzFlypRBa/r6+qKvr6//eW9vb0yfPj16enqipqbmnN8cAACjo7e3N2pra4e110q6ojpx4sSor6+Ptra2AeNtbW3R0NAw5Jr33nsvxo0b+GPGjx8fER9ciR1KVVVV1NTUDHgAADC2lPzRf3Nzczz++OOxadOmOHDgQKxcuTI6Ojr6P8pfvXp1LFu2rH/+bbfdFs8991xs2LAhDh48GK+88kosX7485s+fH1OnTh2+dwIAwEWlstQFS5cujWPHjsW6deuis7Mz5syZE62trTFjxoyIiOjs7BxwT9W77747jh8/Hj/+8Y/j7/7u7+JjH/tY3HDDDfG9731v+N4FAAAXnZK+o3qhjMR3HgAAGD4X/DuqAAAwWoQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAICWhCgBASkIVAICUhCoAACkJVQAAUhKqAACkJFQBAEhJqAIAkJJQBQAgJaEKAEBKQhUAgJSEKgAAKQlVAABSEqoAAKQkVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAICWhCgBASkIVAICUhCoAACkJVQAAUhKqAACkJFQBAEhJqAIAkJJQBQAgJaEKAEBKQhUAgJSEKgAAKQlVAABSEqoAAKQkVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAICWhCgBASkIVAICUhCoAACkJVQAAUhKqAACkJFQBAEhJqAIAkJJQBQAgJaEKAEBKQhUAgJSEKgAAKQlVAABSEqoAAKQkVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAICWhCgBASkIVAICUhCoAACmVFarr16+PmTNnRnV1ddTX18eOHTs+dH5fX1+sWbMmZsyYEVVVVfHJT34yNm3aVNaGAQAYGypLXbBly5ZYsWJFrF+/Pq6//vr4yU9+EosXL479+/fH1VdfPeSa22+/Pd58883YuHFj/Mmf/EkcPXo0Tp48ed6bBwDg4lVRFEVRyoIFCxbE3LlzY8OGDf1js2fPjiVLlkRLS8ug+S+99FJ85StfiYMHD8bll19e1iZ7e3ujtrY2enp6oqampqzXAABg5IxEr5X00f+JEydi79690djYOGC8sbExdu3aNeSaF198MebNmxff//7346qrroprr702Hnjggfjd73531p/T19cXvb29Ax4AAIwtJX30393dHadOnYq6uroB43V1ddHV1TXkmoMHD8bOnTujuro6nn/++eju7o6vfe1r8dZbb531e6otLS2xdu3aUrYGAMBFpqw/pqqoqBjwvCiKQWNnnD59OioqKmLz5s0xf/78uOWWW+Lhhx+OJ5988qxXVVevXh09PT39j8OHD5ezTQAAPsJKuqI6efLkGD9+/KCrp0ePHh10lfWMKVOmxFVXXRW1tbX9Y7Nnz46iKOLIkSNxzTXXDFpTVVUVVVVVpWwNAICLTElXVCdOnBj19fXR1tY2YLytrS0aGhqGXHP99dfHG2+8Ee+8807/2Kuvvhrjxo2LadOmlbFlAADGgpI/+m9ubo7HH388Nm3aFAcOHIiVK1dGR0dHNDU1RcQHH9svW7asf/4dd9wRkyZNinvuuSf2798fL7/8cnzzm9+Mv/mbv4lLLrlk+N4JAAAXlZLvo7p06dI4duxYrFu3Ljo7O2POnDnR2toaM2bMiIiIzs7O6Ojo6J//R3/0R9HW1hZ/+7d/G/PmzYtJkybF7bffHt/97neH710AAHDRKfk+qheC+6gCAOR2we+jCgAAo0WoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAICWhCgBASkIVAICUhCoAACkJVQAAUhKqAACkJFQBAEhJqAIAkJJQBQAgJaEKAEBKQhUAgJSEKgAAKQlVAABSEqoAAKQkVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAICWhCgBASkIVAICUhCoAACkJVQAAUhKqAACkJFQBAEhJqAIAkJJQBQAgJaEKAEBKQhUAgJSEKgAAKQlVAABSEqoAAKQkVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAICWhCgBASkIVAICUhCoAACkJVQAAUhKqAACkJFQBAEhJqAIAkJJQBQAgJaEKAEBKQhUAgJSEKgAAKQlVAABSEqoAAKQkVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBIqaxQXb9+fcycOTOqq6ujvr4+duzYcU7rXnnllaisrIzPfe5z5fxYAADGkJJDdcuWLbFixYpYs2ZNtLe3x6JFi2Lx4sXR0dHxoet6enpi2bJl8Rd/8RdlbxYAgLGjoiiKopQFCxYsiLlz58aGDRv6x2bPnh1LliyJlpaWs677yle+Etdcc02MHz8+Xnjhhdi3b985/8ze3t6ora2Nnp6eqKmpKWW7AACMgpHotZKuqJ44cSL27t0bjY2NA8YbGxtj165dZ133xBNPxGuvvRYPPfTQOf2cvr6+6O3tHfAAAGBsKSlUu7u749SpU1FXVzdgvK6uLrq6uoZc85vf/CZWrVoVmzdvjsrKynP6OS0tLVFbW9v/mD59einbBADgIlDWH1NVVFQMeF4UxaCxiIhTp07FHXfcEWvXro1rr732nF9/9erV0dPT0/84fPhwOdsEAOAj7Nwucf5fkydPjvHjxw+6enr06NFBV1kjIo4fPx579uyJ9vb2+MY3vhEREadPn46iKKKysjK2bdsWN9xww6B1VVVVUVVVVcrWAAC4yJR0RXXixIlRX18fbW1tA8bb2tqioaFh0Pyampr41a9+Ffv27et/NDU1xac+9anYt29fLFiw4Px2DwDARaukK6oREc3NzXHnnXfGvHnzYuHChfHTn/40Ojo6oqmpKSI++Nj+t7/9bfz85z+PcePGxZw5cwasv+KKK6K6unrQOAAA/P9KDtWlS5fGsWPHYt26ddHZ2Rlz5syJ1tbWmDFjRkREdHZ2/sF7qgIAwB9S8n1ULwT3UQUAyO2C30cVAABGi1AFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAICWhCgBASkIVAICUhCoAACkJVQAAUhKqAACkJFQBAEhJqAIAkJJQBQAgJaEKAEBKQhUAgJSEKgAAKQlVAABSEqoAAKQkVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAICWhCgBASkIVAICUhCoAACkJVQAAUhKqAACkJFQBAEhJqAIAkJJQBQAgJaEKAEBKQhUAgJSEKgAAKQlVAABSEqoAAKQkVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAICWhCgBASkIVAICUhCoAACkJVQAAUhKqAACkJFQBAEhJqAIAkJJQBQAgJaEKAEBKQhUAgJSEKgAAKQlVAABSEqoAAKQkVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAIKWyQnX9+vUxc+bMqK6ujvr6+tixY8dZ5z733HNx0003xcc//vGoqamJhQsXxi9+8YuyNwwAwNhQcqhu2bIlVqxYEWvWrIn29vZYtGhRLF68ODo6Ooac//LLL8dNN90Ura2tsXfv3vjiF78Yt912W7S3t5/35gEAuHhVFEVRlLJgwYIFMXfu3NiwYUP/2OzZs2PJkiXR0tJyTq/x2c9+NpYuXRoPPvjgOc3v7e2N2tra6OnpiZqamlK2CwDAKBiJXivpiuqJEydi79690djYOGC8sbExdu3adU6vcfr06Th+/HhcfvnlZ53T19cXvb29Ax4AAIwtJYVqd3d3nDp1Kurq6gaM19XVRVdX1zm9xg9+8IN499134/bbbz/rnJaWlqitre1/TJ8+vZRtAgBwESjrj6kqKioGPC+KYtDYUJ5++un4zne+E1u2bIkrrrjirPNWr14dPT09/Y/Dhw+Xs00AAD7CKkuZPHny5Bg/fvygq6dHjx4ddJX1f9uyZUvce++9sXXr1rjxxhs/dG5VVVVUVVWVsjUAAC4yJV1RnThxYtTX10dbW9uA8ba2tmhoaDjruqeffjruvvvueOqpp+LWW28tb6cAAIwpJV1RjYhobm6OO++8M+bNmxcLFy6Mn/70p9HR0RFNTU0R8cHH9r/97W/j5z//eUR8EKnLli2LH/7wh/H5z3++/2rsJZdcErW1tcP4VgAAuJiUHKpLly6NY8eOxbp166KzszPmzJkTra2tMWPGjIiI6OzsHHBP1Z/85Cdx8uTJ+PrXvx5f//rX+8fvuuuuePLJJ8//HQAAcFEq+T6qF4L7qAIA5HbB76MKAACjRagCAJCSUAUAICWhCgBASkIVAICUhCoAACkJVQAAUhKqAACkJFQBAEhJqAIAkJJQBQAgJaEKAEBKQhUAgJSEKgAAKQlVAABSEqoAAKQkVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAICWhCgBASkIVAICUhCoAACkJVQAAUhKqAACkJFQBAEhJqAIAkJJQBQAgJaEKAEBKQhUAgJSEKgAAKQlVAABSEqoAAKQkVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAICWhCgBASkIVAICUhCoAACkJVQAAUhKqAACkJFQBAEhJqAIAkJJQBQAgJaEKAEBKQhUAgJSEKgAAKQlVAABSEqoAAKQkVAEASEmoAgCQklAFACAloQoAQEpCFQCAlIQqAAApCVUAAFISqgAApCRUAQBISagCAJCSUAUAICWhCgBASkIVAICUhCoAACkJVQAAUhKqAACkJFQBAEiprFBdv359zJw5M6qrq6O+vj527NjxofO3b98e9fX1UV1dHbNmzYrHHnusrM0CADB2lByqW7ZsiRUrVsSaNWuivb09Fi1aFIsXL46Ojo4h5x86dChuueWWWLRoUbS3t8e3v/3tWL58eTz77LPnvXkAAC5eFUVRFKUsWLBgQcydOzc2bNjQPzZ79uxYsmRJtLS0DJr/rW99K1588cU4cOBA/1hTU1P88pe/jN27d5/Tz+zt7Y3a2tro6emJmpqaUrYLAMAoGIleqyxl8okTJ2Lv3r2xatWqAeONjY2xa9euIdfs3r07GhsbB4zdfPPNsXHjxnj//fdjwoQJg9b09fVFX19f//Oenp6I+ODfAAAA8jnTaSVeA/1QJYVqd3d3nDp1Kurq6gaM19XVRVdX15Brurq6hpx/8uTJ6O7ujilTpgxa09LSEmvXrh00Pn369FK2CwDAKDt27FjU1tYOy2uVFKpnVFRUDHheFMWgsT80f6jxM1avXh3Nzc39z99+++2YMWNGdHR0DNsbJ6/e3t6YPn16HD582Fc9xgDnPbY477HFeY8tPT09cfXVV8fll18+bK9ZUqhOnjw5xo8fP+jq6dGjRwddNT3jyiuvHHJ+ZWVlTJo0acg1VVVVUVVVNWi8trbWf9DHkJqaGuc9hjjvscV5jy3Oe2wZN2747n5a0itNnDgx6uvro62tbcB4W1tbNDQ0DLlm4cKFg+Zv27Yt5s2bN+T3UwEAIKKM21M1NzfH448/Hps2bYoDBw7EypUro6OjI5qamiLig4/tly1b1j+/qakpXn/99Whubo4DBw7Epk2bYuPGjfHAAw8M37sAAOCiU/J3VJcuXRrHjh2LdevWRWdnZ8yZMydaW1tjxowZERHR2dk54J6qM2fOjNbW1li5cmU8+uijMXXq1HjkkUfiy1/+8jn/zKqqqnjooYeG/DoAFx/nPbY477HFeY8tzntsGYnzLvk+qgAAMBqG79uuAAAwjIQqAAApCVUAAFISqgAApJQmVNevXx8zZ86M6urqqK+vjx07dnzo/O3bt0d9fX1UV1fHrFmz4rHHHhulnTIcSjnv5557Lm666ab4+Mc/HjU1NbFw4cL4xS9+MYq75XyV+vt9xiuvvBKVlZXxuc99bmQ3yLAq9bz7+vpizZo1MWPGjKiqqopPfvKTsWnTplHaLeer1PPevHlzXHfddXHppZfGlClT4p577oljx46N0m4p18svvxy33XZbTJ06NSoqKuKFF174g2uGpdWKBP75n/+5mDBhQvGzn/2s2L9/f3H//fcXl112WfH6668POf/gwYPFpZdeWtx///3F/v37i5/97GfFhAkTimeeeWaUd045Sj3v+++/v/je975X/Od//mfx6quvFqtXry4mTJhQ/Pd///co75xylHreZ7z99tvFrFmzisbGxuK6664bnc1y3so57y996UvFggULira2tuLQoUPFf/zHfxSvvPLKKO6acpV63jt27CjGjRtX/PCHPywOHjxY7Nixo/jsZz9bLFmyZJR3TqlaW1uLNWvWFM8++2wREcXzzz//ofOHq9VShOr8+fOLpqamAWOf/vSni1WrVg05/+///u+LT3/60wPGvvrVrxaf//znR2yPDJ9Sz3son/nMZ4q1a9cO99YYAeWe99KlS4t/+Id/KB566CGh+hFS6nn/y7/8S1FbW1scO3ZsNLbHMCv1vP/xH/+xmDVr1oCxRx55pJg2bdqI7ZHhdy6hOlytdsE/+j9x4kTs3bs3GhsbB4w3NjbGrl27hlyze/fuQfNvvvnm2LNnT7z//vsjtlfOXznn/b+dPn06jh8/HpdffvlIbJFhVO55P/HEE/Haa6/FQw89NNJbZBiVc94vvvhizJs3L77//e/HVVddFddee2088MAD8bvf/W40tsx5KOe8Gxoa4siRI9Ha2hpFUcSbb74ZzzzzTNx6662jsWVG0XC1Wsn/z1TDrbu7O06dOhV1dXUDxuvq6qKrq2vINV1dXUPOP3nyZHR3d8eUKVNGbL+cn3LO+3/7wQ9+EO+++27cfvvtI7FFhlE55/2b3/wmVq1aFTt27IjKygv+jyhKUM55Hzx4MHbu3BnV1dXx/PPPR3d3d3zta1+Lt956y/dUkyvnvBsaGmLz5s2xdOnS+P3vfx8nT56ML33pS/GjH/1oNLbMKBquVrvgV1TPqKioGPC8KIpBY39o/lDj5FTqeZ/x9NNPx3e+853YsmVLXHHFFSO1PYbZuZ73qVOn4o477oi1a9fGtddeO1rbY5iV8vt9+vTpqKioiM2bN8f8+fPjlltuiYcffjiefPJJV1U/Iko57/3798fy5cvjwQcfjL1798ZLL70Uhw4diqamptHYKqNsOFrtgl+umDx5cowfP37Q//o6evTooBI/48orrxxyfmVlZUyaNGnE9sr5K+e8z9iyZUvce++9sXXr1rjxxhtHcpsMk1LP+/jx47Fnz55ob2+Pb3zjGxHxQcgURRGVlZWxbdu2uOGGG0Zl75SunN/vKVOmxFVXXRW1tbX9Y7Nnz46iKOLIkSNxzTXXjOieKV85593S0hLXX399fPOb34yIiD/90z+Nyy67LBYtWhTf/e53fSJ6ERmuVrvgV1QnTpwY9fX10dbWNmC8ra0tGhoahlyzcOHCQfO3bdsW8+bNiwkTJozYXjl/5Zx3xAdXUu++++546qmnfJfpI6TU866pqYlf/epXsW/fvv5HU1NTfOpTn4p9+/bFggULRmvrlKGc3+/rr78+3njjjXjnnXf6x1599dUYN25cTJs2bUT3y/kp57zfe++9GDduYHqMHz8+Iv7f1TYuDsPWaiX96dUIOXN7i40bNxb79+8vVqxYUVx22WXF//zP/xRFURSrVq0q7rzzzv75Z255sHLlymL//v3Fxo0b3Z7qI6TU837qqaeKysrK4tFHHy06Ozv7H2+//faFeguUoNTz/t/81f9HS6nnffz48WLatGnFX/3VXxW//vWvi+3btxfXXHNNcd99912ot0AJSj3vJ554oqisrCzWr19fvPbaa8XOnTuLefPmFfPnz79Qb4FzdPz48aK9vb1ob28vIqJ4+OGHi/b29v5bkY1Uq6UI1aIoikcffbSYMWNGMXHixGLu3LnF9u3b+/+1u+66q/jCF74wYP6//du/FX/2Z39WTJw4sfjEJz5RbNiwYZR3zPko5by/8IUvFBEx6HHXXXeN/sYpS6m/3/8/ofrRU+p5HzhwoLjxxhuLSy65pJg2bVrR3NxcvPfee6O8a8pV6nk/8sgjxWc+85nikksuKaZMmVL89V//dXHkyJFR3jWl+td//dcP/e/ikWq1iqJwrR0AgHwu+HdUAQBgKEIVAICUhCoAACkJVQAAUhKqAACkJFQBAEhJqAIAkJJQBQAgJaEKAEBKQhUAgJSEKgAAKQlVAABS+j8pDd1lCZ/1kAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 800x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#initialize UMAP reducer \n",
    "umap_reducer = umap.UMAP(n_neighbors=15, min_dist=0.1, metric='euclidean')\n",
    "\n",
    "# Apply UMAP\n",
    "umap_embedding = umap.UMAP(n_neighbors=15, min_dist=0.1, metric='euclidean').fit_transform(umap_data.iloc[1:, :])\n",
    "\n",
    "# Convert lineage labels to categorical codes for coloring\n",
    "colors = pd.Categorical(lineage_labels).codes\n",
    "\n",
    "# UMAP plot\n",
    "plt.figure(figsize=(8, 6))\n",
    "plt.scatter(umap_embedding[:, 0], umap_embedding[:, 1], c=colors, cmap='Spectral', s=5, alpha=0.8)\n",
    "plt.title('UMAP of ATAC Samples by Cell Lineage')\n",
    "plt.xlabel('UMAP1')\n",
    "plt.ylabel('UMAP2')\n",
    "plt.colorbar(label='Cell Lineage')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f468414",
   "metadata": {},
   "source": [
    "## Create UMAP dimensionality reduction to check for outlier cells\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "43a653bf",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'cell_lineages'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/dataanalysisenv/lib/python3.13/site-packages/pandas/core/indexes/base.py:3805\u001b[39m, in \u001b[36mIndex.get_loc\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   3804\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m-> \u001b[39m\u001b[32m3805\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_engine\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   3806\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "\u001b[36mFile \u001b[39m\u001b[32mindex.pyx:167\u001b[39m, in \u001b[36mpandas._libs.index.IndexEngine.get_loc\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mindex.pyx:196\u001b[39m, in \u001b[36mpandas._libs.index.IndexEngine.get_loc\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[39m, in \u001b[36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[39m, in \u001b[36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[31mKeyError\u001b[39m: 'cell_lineages'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[7]\u001b[39m\u001b[32m, line 9\u001b[39m\n\u001b[32m      5\u001b[39m umap_embedding = umap_reducer.fit_transform(ATAC_scores.iloc[:, \u001b[32m1\u001b[39m:])\n\u001b[32m      7\u001b[39m \u001b[38;5;66;03m# Plot results\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m9\u001b[39m plt.scatter(umap_embedding[:, \u001b[32m0\u001b[39m], umap_embedding[:, \u001b[32m1\u001b[39m], c=\u001b[43mATAC_scores\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mcell_lineages\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m]\u001b[49m, cmap=\u001b[33m'\u001b[39m\u001b[33mSpectral\u001b[39m\u001b[33m'\u001b[39m, s=\u001b[32m5\u001b[39m, alpha=\u001b[32m0.7\u001b[39m)\n\u001b[32m     10\u001b[39m plt.title(\u001b[33m'\u001b[39m\u001b[33mUMAP of ATAC Scores, colored by cell type\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m     11\u001b[39m plt.xlabel(\u001b[33m'\u001b[39m\u001b[33mUMAP1\u001b[39m\u001b[33m'\u001b[39m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/dataanalysisenv/lib/python3.13/site-packages/pandas/core/frame.py:4102\u001b[39m, in \u001b[36mDataFrame.__getitem__\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   4100\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.columns.nlevels > \u001b[32m1\u001b[39m:\n\u001b[32m   4101\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m._getitem_multilevel(key)\n\u001b[32m-> \u001b[39m\u001b[32m4102\u001b[39m indexer = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mcolumns\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   4103\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[32m   4104\u001b[39m     indexer = [indexer]\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/miniconda3/envs/dataanalysisenv/lib/python3.13/site-packages/pandas/core/indexes/base.py:3812\u001b[39m, in \u001b[36mIndex.get_loc\u001b[39m\u001b[34m(self, key)\u001b[39m\n\u001b[32m   3807\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[32m   3808\u001b[39m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc.Iterable)\n\u001b[32m   3809\u001b[39m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[32m   3810\u001b[39m     ):\n\u001b[32m   3811\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[32m-> \u001b[39m\u001b[32m3812\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01merr\u001b[39;00m\n\u001b[32m   3813\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[32m   3814\u001b[39m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[32m   3815\u001b[39m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[32m   3816\u001b[39m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[32m   3817\u001b[39m     \u001b[38;5;28mself\u001b[39m._check_indexing_error(key)\n",
      "\u001b[31mKeyError\u001b[39m: 'cell_lineages'"
     ]
    }
   ],
   "source": [
    "#initialize UMAP reducer \n",
    "umap_reducer = umap.UMAP(n_neighbors=15, min_dist=0.1, metric='euclidean')\n",
    "\n",
    "# Fit and transform the ATAC scores using UMAP\n",
    "umap_embedding = umap_reducer.fit_transform(ATAC_scores.iloc[:, 1:])\n",
    "\n",
    "# Plot results\n",
    "\n",
    "plt.scatter(umap_embedding[:, 0], umap_embedding[:, 1], c=ATAC_scores['cell_lineages'], cmap='Spectral', s=5, alpha=0.7)\n",
    "plt.title('UMAP of ATAC Scores, colored by cell type')\n",
    "plt.xlabel('UMAP1')\n",
    "plt.ylabel('UMAP2')\n",
    "plt.show()\n",
    "\n",
    "#create a DataFrame with UMAP coordinates\n",
    "umap_df = pd.DataFrame(umap_embedding, columns=['UMAP1', 'UMAP2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "29279ccf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['chrom', 'Summit', 'mm10.60way.phastCons_scores', '_-log10_bestPvalue',\n",
      "       'Included.in.systematic.analysis', 'TSS', 'genes.within.100Kb',\n",
      "       'LTHSC.34-.BM', 'LTHSC.34+.BM', 'STHSC.150-.BM', 'MPP4.135+.BM',\n",
      "       'proB.CLP.BM', 'proB.FrA.BM', 'proB.FrBC.BM', 'preB.FrD.BM', 'B.FrE.BM',\n",
      "       'B1b.PC', 'B.T1.Sp', 'B.T2.Sp', 'B.T3.Sp', 'B.Sp', 'B.Fem.Sp',\n",
      "       'B.MZ.Sp', 'B.Fo.Sp', 'B.mem.Sp', 'B.GC.CB.Sp', 'B.GC.CC.Sp', 'B.PB.Sp',\n",
      "       'B.PC.Sp', 'B.PC.BM', 'preT.DN1.Th', 'preT.DN2a.Th', 'preT.DN2b.Th',\n",
      "       'preT.DN3.Th', 'T.DN4.Th', 'T.ISP.Th', 'T.DP.Th', 'T.4.Th', 'T.8.Th',\n",
      "       'T.4.Nve.Sp', 'T.4.Nve.Fem.Sp', 'T.8.Nve.Sp', 'T.4.Sp.aCD3+CD40.18hr',\n",
      "       'Treg.4.FP3+.Nrplo.Co', 'Treg.4.25hi.Sp', 'T8.TN.P14.Sp',\n",
      "       'T8.IEL.LCMV.d7.SI', 'T8.TE.LCMV.d7.Sp', 'T8.MP.LCMV.d7.Sp',\n",
      "       'T8.Tcm.LCMV.d180.Sp', 'T8.Tem.LCMV.d180.Sp', 'NKT.Sp',\n",
      "       'NKT.Sp.LPS.3hr', 'NKT.Sp.LPS.18hr', 'NKT.Sp.LPS.3d',\n",
      "       'Tgd.g2+d17.24a+.Th', 'Tgd.g2+d17.LN', 'Tgd.g2+d1.24a+.Th',\n",
      "       'Tgd.g2+d1.LN', 'Tgd.g1.1+d1.24a+.Th', 'Tgd.g1.1+d1.LN', 'Tgd.Sp',\n",
      "       'NK.27+11b-.BM', 'NK.27+11b+.BM', 'NK.27-11b+.BM', 'NK.27+11b-.Sp',\n",
      "       'NK.27+11b+.Sp', 'NK.27-11b+.Sp', 'ILC2.SI', 'ILC3.NKp46-CCR6-.SI',\n",
      "       'ILC3.NKp46+.SI', 'ILC3.CCR6+.SI', 'MPP3.48+.BM', 'GN.BM', 'GN.Sp',\n",
      "       'GN.Thio.PC', 'Mo.6C+II-.Bl', 'Mo.6C-II-.Bl', 'MF.pIC.Alv.Lu',\n",
      "       'MF.226+II+480lo.PC', 'MF.ICAM+480hi.PC', 'MF.RP.Sp', 'MF.LP.SI',\n",
      "       'MF.Alv.Lu', 'MF.microglia.CNS', 'MF.PC', 'MF.Fem.PC', 'DC.4+.Sp',\n",
      "       'DC.8+.Sp', 'DC.pDC.Sp', 'DC.103+11b+.SI', 'DC.103+11b-.SI', 'FRC.SLN',\n",
      "       'IAP.SLN', 'BEC.SLN', 'LEC.SLN', 'Ep.MEChi.Th'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(ATAC_cleaned.columns[1:])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e20e6a76",
   "metadata": {},
   "source": [
    "UMAP for subset of cell types relevant for ab gd T cell differentiation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d955243",
   "metadata": {},
   "outputs": [],
   "source": [
    "# columns sorted by organs\n",
    "\n",
    "bm_cols = ['LTHSC.34-.BM', 'LTHSC.34+.BM', 'STHSC.150-.BM', 'MPP4.135+.BM', 'proB.CLP.BM', 'proB.FrA.BM', 'proB.FrBC.BM', 'preB.FrD.BM', \n",
    "        'B.FrE.BM', 'NK.27+11b-.BM', 'NK.27+11b+.BM', 'NK.27-11b+.BM', 'MPP3.48+.BM', 'GN.BM']\n",
    "\n",
    "sp_cols = ['B.T1.Sp', 'B.T2.Sp', 'B.T3.Sp', 'B.Sp', 'B.Fem.Sp',\n",
    "       'B.MZ.Sp', 'B.Fo.Sp', 'B.mem.Sp', 'B.GC.CB.Sp', 'B.GC.CC.Sp', 'B.PB.Sp',\n",
    "       'B.PC.Sp', 'T.4.Nve.Sp', 'T.4.Nve.Fem.Sp', 'T.8.Nve.Sp', 'T.4.Sp.aCD3+CD40.18hr', 'Treg.4.25hi.Sp', 'T8.TN.P14.Sp',\n",
    "       'T8.IEL.LCMV.d7.SI', 'T8.TE.LCMV.d7.Sp', 'T8.MP.LCMV.d7.Sp',\n",
    "       'T8.Tcm.LCMV.d180.Sp', 'T8.Tem.LCMV.d180.Sp', 'NKT.Sp',\n",
    "       'NKT.Sp.LPS.3hr', 'NKT.Sp.LPS.18hr', 'NKT.Sp.LPS.3d', 'Tgd.Sp', 'NK.27+11b-.Sp',\n",
    "       'NK.27+11b+.Sp', 'NK.27-11b+.Sp', 'GN.Sp', 'MF.RP.Sp', 'DC.4+.Sp',\n",
    "       'DC.8+.Sp', 'DC.pDC.Sp']\n",
    "\n",
    "thy_cols = ['preT.DN1.Th', 'preT.DN2a.Th', 'preT.DN2b.Th', 'preT.DN3.Th', 'T.DN4.Th', 'T.ISP.Th', 'T.DP.Th', 'T.4.Th', 'T.8.Th', \n",
    "        'Tgd.g2+d17.24a+.Th', 'Tgd.g2+d1.24a+.Th', 'Tgd.g1.1+d1.24a+.Th', 'Ep.MEChi.Th']\n",
    "\n",
    "col_cols = ['Treg.4.FP3+.Nrplo.Co']\n",
    "\n",
    "small_int_cols = ['DC.103+11b+.SI', 'DC.103+11b-.SI', 'MF.LP.SI', 'ILC2.SI', 'ILC3.NKp46-CCR6-.SI', 'ILC3.NKp46+.SI', 'ILC3.CCR6+.SI',\n",
    "        'MF.LP.SI', 'DC.103+11b+.SI', 'DC.103+11b-.SI']\n",
    "\n",
    "sk_dr_LN_cols = ['FRC.SLN', 'IAP.SLN', 'BEC.SLN', 'LEC.SLN']\n",
    "\n",
    "per_cav_cols = ['B1b.PC', 'GN.Thio.PC', 'MF.226+II+480lo.PC', 'MF.ICAM+480hi.PC', 'MF.PC', 'MF.Fem.PC']\n",
    "\n",
    "lung_cols = ['MF.Alv.Lu', 'MF.pIC.Alv.Lu']\n",
    "\n",
    "per_bl_cols = ['Mo.6C+II-.Bl', 'Mo.6C-II-.Bl']\n",
    "\n",
    "ln_cols = ['Tgd.g2+d17.LN', 'Tgd.g2+d1.LN', 'Tgd.g1.1+d1.LN']\n",
    "\n",
    "brain_cols = ['MF.microglia.CNS']\n",
    "\n",
    "organs = {\n",
    "\n",
    "    'BM': bm_cols,\n",
    "    'SP': sp_cols,\n",
    "    'Thy': thy_cols,\n",
    "    'Col': col_cols,\n",
    "    'Small_int': small_int_cols,\n",
    "    'Sk_dr_LN': sk_dr_LN_cols,\n",
    "    'Per_cav': per_cav_cols,\n",
    "    'Lung': lung_cols,\n",
    "    'Per_bl': per_bl_cols,\n",
    "    'LN': ln_cols,\n",
    "    'Brain': brain_cols\n",
    "    }"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dataanalysisenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
